<!DOCTYPE html>
<html prefix="
og: http://ogp.me/ns# article: http://ogp.me/ns/article#
" lang="en">
<head>
<meta charset="utf-8">
<meta name="description" content="Metagenomics Classification - Workshop material">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Metagenomics Machine Learning | Avil Page</title>
<link href="assets/css/bootstrap.min.css" rel="stylesheet" type="text/css">
<link href="assets/css/baguetteBox.min.css" rel="stylesheet" type="text/css">
<link href="assets/css/rst.css" rel="stylesheet" type="text/css">
<link href="assets/css/code.css" rel="stylesheet" type="text/css">
<link href="assets/css/theme.css" rel="stylesheet" type="text/css">
<meta name="theme-color" content="#5670d4">
<meta name="generator" content="Nikola (getnikola.com)">
<link rel="alternate" type="application/rss+xml" title="RSS" hreflang="en" href="rss.xml">
<link rel="canonical" href="https://avilpage.com/metagenomics-machine-learning.html">
<!--[if lt IE 9]><script src="assets/js/html5.js"></script><![endif]--><meta name="author" content="Anand Reddy Pandikunta">
<meta property="og:site_name" content="Avil Page">
<meta property="og:title" content="Metagenomics Machine Learning">
<meta property="og:url" content="https://avilpage.com/metagenomics-machine-learning.html">
<meta property="og:description" content="Metagenomics Classification - Workshop material">
<meta property="og:type" content="article">
<meta property="article:published_time" content="2024-08-08T03:29:59Z">
<meta property="article:tag" content="bioinformatics">
<meta property="article:tag" content="metagenomics">
<style>
    .full-article-footer {
    border-top: 1px solid #E0DFDB;
}


.article-footer {
    padding: 20px 10px 10px 10px;
    min-height: 115px;
    display:table;
}


.avatar-module {
    display: table-cell;
    vertical-align: middle;
}


.avatar-module img {
    border-radius: 50%!important;
    height: 120px;
    width: 120px;
    float: left;
    margin:0px 20px 0px 20px;
}


.article-footer p {
    line-height:1.5em;
    padding-left:15px;
}


@media (max-width:750px) {
    .article-footer {
        display:inherit;
        margin:0px;
    }

    .avatar-module {
        display:inherit;
        vertical-align: none;
    }

    .avatar-module img {
        margin-left: auto;
        margin-right: auto;
        display: block;
        float:none;
        margin-top:0px;
    }

    .article-footer p {
        text-align:center;
        padding-left:0px;
    }
}

    </style>
<style>
@font-face {
    font-family: 'LovelexieHandwritten';
    src: url('files/LovelexieHandwritten.ttf') format('truetype');
    font-weight: normal;
    font-style: normal;
}

h1,h2,h3,h4,h5,h6 {
    font-family: 'LovelexieHandwritten', serif;
}

 h1 {
     font-weight: bold;
     font-size: 25px;
 }

</style>
</head>
<body>
    <a href="#content" class="sr-only sr-only-focusable">Skip to main content</a>

    <!-- Menubar -->
    <nav class="navbar navbar-expand-md static-top mb-4
                navbar-dark
                bg-dark
                "><div class="container">
<!-- This keeps the margins nice -->
            <a class="navbar-brand" href=".">

                    <span id="blog-title">Avil Page</span>
            </a>
            <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#bs-navbar" aria-controls="bs-navbar" aria-expanded="false" aria-label="Toggle navigation">
                <span class="navbar-toggler-icon"></span>
            </button>

            <div class="collapse navbar-collapse" id="bs-navbar">
                <ul class="navbar-nav mr-auto">
<li class="nav-item">
<a href="guides.html" class="nav-link">Guides</a>
                </li>
<li class="nav-item">
<a href="archive.html" class="nav-link">Blog</a>
                </li>
<li class="nav-item">
<a href="books.html" class="nav-link">Books</a>
                </li>
<li class="nav-item">
<a href="talks.html" class="nav-link">Talks</a>
                </li>
<li class="nav-item">
<a href="projects.html" class="nav-link">Projects</a>
                </li>
<li class="nav-item">
<a href="rss.xml" class="nav-link">RSS</a>

                    
                </li>
</ul>
<ul class="navbar-nav navbar-right"></ul>
</div>
<!-- /.navbar-collapse -->
        </div>
<!-- /.container -->
    </nav><!-- End of Menubar --><div class="container" id="content" role="main">
    <div class="row justify-content-md-center">
        <div class="body-content col col-lg-6">
            <!--Body content-->
            
<article class="post-text storypage" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title" itemprop="headline name"><a href="#" class="u-url">Metagenomics Machine Learning</a></h1>

        
    </header><div class="e-content entry-content" itemprop="articleBody text">
    <h4>Introduction</h4>
<p>This is the reference material for the "Machine Learning in Metagenomics" workshop. By the end of workshop, you will be able to classify metagenomic samples using Kraken2 and predict the source organism of the sample with simple ML methods.</p>
<p><a href="metagenomics-deep-learning.html">Click here for "Metagenomics - Deep Learning" workshop</a></p>
<h4>Environment Setup</h4>
<p>We will be using Python, Jupyter notebook and command line tools for this workshop.</p>
<p>If you are using Windows, you can install WSL2 and use Ubuntu. If you are using Mac, you can use the terminal.</p>
<h4>Running Kraken2</h4>
<p>In the first step, let's generate abundance report using Kraken2. Kraken2 is a taxonomic sequence classifier that assigns taxonomic labels to short DNA sequences.</p>
<p>Install anaconda</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>wget<span class="w"> </span>https://repo.anaconda.com/archive/Anaconda3-2024.06-1-Linux-x86_64.sh
$<span class="w"> </span>bash<span class="w"> </span>Anaconda3-2021.05-Linux-x86_64.sh
</pre></div>

<p>Install kraken2</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>conda<span class="w"> </span>install<span class="w"> </span>-c<span class="w"> </span>bioconda<span class="w"> </span>kraken2
</pre></div>

<p>Download pre-built viral database</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>wget<span class="w"> </span>-c<span class="w"> </span>https://genome-idx.s3.amazonaws.com/kraken/k2_viral_20240605.tar.gz
$<span class="w"> </span>mkdir<span class="w"> </span>-p<span class="w"> </span>k2_viral
$<span class="w"> </span>tar<span class="w"> </span>-xzvf<span class="w"> </span>k2_viral_20240605.tar.gz<span class="w"> </span>-C<span class="w"> </span>k2_viral
</pre></div>

<p>Download sample fastq files</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>wget<span class="w"> </span>-c<span class="w"> </span>https://ftp.sra.ebi.ac.uk/vol1/fastq/ERR103/077/ERR10359977/ERR10359977.fastq.gz
</pre></div>

<p>Kraken2 classification</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>kraken2<span class="w"> </span>--db<span class="w"> </span>k2_viral<span class="w"> </span>ERR10359977.fastq.gz<span class="w"> </span>--report<span class="w"> </span>k2_report.txt<span class="w"> </span>--output<span class="w"> </span>k2_output.txt
</pre></div>

<h4>Generate Krona plot</h4>
<p>Install Krona &amp; update taxonomy</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>conda<span class="w"> </span>install<span class="w"> </span>-c<span class="w"> </span>bioconda<span class="w"> </span>krona
$<span class="w"> </span>ktUpdateTaxonomy.sh
</pre></div>

<p>Generate Krona plot</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>ktImportTaxonomy<span class="w"> </span>k2_report.txt<span class="w"> </span>-o<span class="w"> </span>k2_report.html
</pre></div>

<h4>Building custom database</h4>
<p>Download taxanomy</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>kraken2-build<span class="w"> </span>--download-taxonomy<span class="w"> </span>--db<span class="w"> </span>k2_fungi
</pre></div>

<p>Download required genomes</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>ncbi-genome-download<span class="w"> </span>--format<span class="w"> </span>fasta<span class="w"> </span>--section<span class="w"> </span>refseq<span class="w"> </span>--assembly-level<span class="w"> </span><span class="nb">complete</span><span class="w"> </span>fungi<span class="w"> </span>-v
</pre></div>

<p>Build kraken2 database</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>find<span class="w"> </span>refseq<span class="w"> </span>-name<span class="w"> </span><span class="s2">"*.gz"</span><span class="w"> </span>-print0<span class="w"> </span><span class="p">|</span><span class="w"> </span>parallel<span class="w"> </span>-0<span class="w"> </span>gunzip

$<span class="w"> </span>find<span class="w"> </span>refseq<span class="w"> </span>-name<span class="w"> </span><span class="s2">"*.fna"</span><span class="w"> </span>-exec<span class="w"> </span>kraken2-build<span class="w"> </span>--add-to-library<span class="w"> </span><span class="o">{}</span><span class="w"> </span>--db<span class="w"> </span>custom_db<span class="w"> </span><span class="se">\;</span>

$<span class="w"> </span>kraken2-build<span class="w"> </span>--db<span class="w"> </span>custom_db<span class="w"> </span>--build<span class="w"> </span>--threads<span class="w"> </span><span class="m">36</span>
</pre></div>

<p>Using <code>kraken-db-builder</code> to build database</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>pip<span class="w"> </span>install<span class="w"> </span>ncbi-genome-download<span class="w"> </span>kraken-db-builder

$<span class="w"> </span>kraken-db-builder<span class="w"> </span>--db-type<span class="w"> </span>fungi
</pre></div>

<h4>Predicting Source of Metagenomic Sample</h4>
<p>Geometry Mean of Pairwise Ratios (GMPR)</p>
<p>Bray-Curtis dissimilarity</p>
<p>t-Stochastic Neighbor Embedding (t-SNE)</p>
<p>k-Nearest Neighbors (k-NN)</p>
<p>Download sample data</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>wget<span class="w"> </span>https://raw.githubusercontent.com/maxibor/sourcepredict/master/data/test/dog_test_sink_sample.csv<span class="w"> </span>-O<span class="w"> </span>dog_example.csv
$<span class="w"> </span>wget<span class="w"> </span>https://raw.githubusercontent.com/maxibor/sourcepredict/master/data/modern_gut_microbiomes_labels.csv<span class="w"> </span>-O<span class="w"> </span>sp_labels.csv
$<span class="w"> </span>wget<span class="w"> </span>https://raw.githubusercontent.com/maxibor/sourcepredict/master/data/modern_gut_microbiomes_sources.csv<span class="w"> </span>-O<span class="w"> </span>sp_sources.csv
</pre></div>

<p>Download sourcepredict</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>python<span class="w"> </span>-m<span class="w"> </span>pip<span class="w"> </span>install<span class="w"> </span>git+https://github.com/AvilPage/sourcepredict
</pre></div>

<h4>Machine Learning Notebooks</h4>
<p><a href="https://github.com/ChillarAnand/avilpage.com/tree/master/mg_workshop">https://github.com/ChillarAnand/avilpage.com/tree/master/mg_workshop</a></p>
<h4>Useful links:</h4>
<p><a href="https://benlangmead.github.io/aws-indexes/k2">https://benlangmead.github.io/aws-indexes/k2</a></p>
<p><a href="https://jszym.com/blog/dna_protein_complexity">https://jszym.com/blog/dna_protein_complexity/</a></p>
<p><a href="https://avilpage.com/2024/07/mastering-kraken2-initial-runs.html">https://avilpage.com/2024/07/mastering-kraken2-initial-runs.html</a></p>
<p><a href="https://github.com/maxibor/sourcepredict">https://github.com/maxibor/sourcepredict</a></p>
    </div>
    
</article><!--End of body content--><footer id="footer"><div class="container align-items-center justify-content-center d-flex">

<footer class="footer"><a href="https://github.com/ChillarAnand">
<img src="images/icons8-github.svg" alt="github-chillar-anand" height="34"></a>

  

<a href="https://stackoverflow.com/users/2698552/chillar-anand">
<img src="images/icons8-so.svg" alt="github-chillar-anand" height="30"></a>

  

<a href="https://youtube.com/@avilpage">
<img src="images/icons8-youtube.svg" alt="github-chillar-anand" height="34"></a>

  

<a href="https://linkedin.com/in/chillaranand">
<img src="images/icons8-linkedin.svg" alt="github-chillar-anand" height="34"></a>

  

<a href="https://twitter.com/chillaranand">
<img src="images/icons8-twitter.svg" alt="github-chillar-anand" height="34"></a>

</footer>
</div>

<br><script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-8127755709093749" crossorigin="anonymous"></script></footer>
</div>
    </div>
    </div>

                <script src="assets/js/jquery.min.js"></script><script src="assets/js/popper.min.js"></script><script src="assets/js/bootstrap.min.js"></script><script src="assets/js/baguetteBox.min.js"></script><script>
    baguetteBox.run('div#content', {
        ignoreClass: 'islink',
        captions: function(element){var i=element.getElementsByTagName('img')[0];return i===undefined?'':i.alt;}});
    </script>
</body>
</html>
