<html><body><div dir="ltr" style="text-align: left;" trbidi="on"><div class="separator" style="clear: both; text-align: center;"></div><div class="separator" style="clear: both; text-align: center;"></div>FASTQ has become the de facto standard for storing the output of high throughput sequencing instruments such as the Illumina Genome Analyzer. So, it is necessary for bioinformatics scientists to zip/unzip, read &amp; process these files on daily basis. This is a simple tutorial to get started working with those files.<br><br><h3 style="text-align: left;">1. Understand Structure Of FASTQ File.</h3>FASTQ format is a text-based format for storing both a biological sequence (usually nucleotide sequence) and its corresponding quality scores. Both the sequence letter and quality score are encoded with a single ASCII character for brevity. A FASTQ file normally uses four lines per sequence.<br><br>Lets look at first 4 lines of sample FASTQ file.<br><br><div class="separator" style="clear: both; text-align: center;"><a href="http://2.bp.blogspot.com/-Dz3ilfAmpwc/U6QxO8t0TrI/AAAAAAAAKko/2UAqomwFtIY/s1600/read-zip-fastq-file.png" imageanchor="1" style="margin-left: 1em; margin-right: 1em;"><img border="0" src="http://2.bp.blogspot.com/-Dz3ilfAmpwc/U6QxO8t0TrI/AAAAAAAAKko/2UAqomwFtIY/s1600/read-zip-fastq-file.png"></a></div><br><ul style="text-align: left;"><li>Line 1 begins with a '@' character and is followed by a sequence identifier and an optional description (like a FASTA title line). </li><li>Line 2 is the raw sequence letters. </li><li>Line 3 begins with a '+' character and is optionally followed by the same sequence identifier (and any description) again. </li><li>Line 4 encodes the quality values for the sequence in Line 2, and must contain the same number of symbols as letters in the sequence.</li></ul><div><h3 style="text-align: left;">2. Find Number Of Lines.</h3></div><div>Usually the size of FASTQ file will be few GB and most of them will be compressed. So, it is difficult to extract such a file. But don't worry, Linux can handle those files quite easily without even extracting. To find the number of lines in a file just use </div><div>  <br>                                          bzcat filename | wc -l<br><br></div><div>to get number of lines in the file. Divinding that by 4 will give you number of reads in that file.<br><br><h3 style="text-align: left;">3. Split File Into Chunks:</h3><div>Most of the times it is necessary to split the huge file into several chunks. </div><div>                                   <br>                                  split -l 400000 sample.fastq sample_part<br><br></div><div>will split the the file into several parts each containing 400,000 lines and their names start with parts.<br><br><br><div style="-webkit-text-stroke-width: 0px; color: black; font-family: 'Times New Roman'; font-size: medium; font-style: normal; font-variant: normal; font-weight: normal; letter-spacing: normal; line-height: normal; orphans: auto; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px;"></div><br><h3 style="-webkit-text-stroke-width: 0px; color: black; font-family: 'Times New Roman'; font-style: normal; font-variant: normal; letter-spacing: normal; line-height: normal; orphans: auto; text-align: left; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px;">4. Unzipping The File: </h3></div><div>Also sometimes it is necessary to unzip the file for processing and analyzing. But if you normal unzip commands it will throw an error. You should use<br><br><br></div><div>                      zcat myreads.fastq.gz | split -d -l 2000000 - Block</div><div><br></div><div><br></div></div></div></body></html>