<!DOCTYPE html>
<html prefix="
og: http://ogp.me/ns# article: http://ogp.me/ns/article#
" lang="en">
<head>
<meta charset="utf-8">
<meta name="description" content="How to speed up kraken2 classification process">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Mastering Kraken2 - Part 1 - Initial Runs | Avil Page</title>
<link href="../../assets/css/bootstrap.min.css" rel="stylesheet" type="text/css">
<link href="../../assets/css/baguetteBox.min.css" rel="stylesheet" type="text/css">
<link href="../../assets/css/rst.css" rel="stylesheet" type="text/css">
<link href="../../assets/css/code.css" rel="stylesheet" type="text/css">
<link href="../../assets/css/theme.css" rel="stylesheet" type="text/css">
<meta name="theme-color" content="#5670d4">
<meta name="generator" content="Nikola (getnikola.com)">
<link rel="alternate" type="application/rss+xml" title="RSS" hreflang="en" href="../../rss.xml">
<link rel="canonical" href="https://avilpage.com/2024/07/mastering-kraken2-initial-runs.html">
<!--[if lt IE 9]><script src="../../assets/js/html5.js"></script><![endif]--><meta name="author" content="Chillar Anand">
<link rel="prev" href="../06/headlamp-k8s-lens-open-source-alternative.html" title="Headlamp - k8s Lens open source alternative" type="text/html">
<link rel="next" href="mastering-kraken2-performance-optimisation.html" title="Mastering kraken2 - Part 2 - Performance Optimisation" type="text/html">
<meta property="og:site_name" content="Avil Page">
<meta property="og:title" content="Mastering Kraken2 - Part 1 - Initial Runs">
<meta property="og:url" content="https://avilpage.com/2024/07/mastering-kraken2-initial-runs.html">
<meta property="og:description" content="How to speed up kraken2 classification process">
<meta property="og:type" content="article">
<meta property="article:published_time" content="2024-07-28T10:44:25+05:30">
<meta property="article:tag" content="devops">
<meta property="article:tag" content="kraken2">
<meta property="article:tag" content="metagenomics">
<style>
    .full-article-footer {
    border-top: 1px solid #E0DFDB;
}


.article-footer {
    padding: 20px 10px 10px 10px;
    min-height: 115px;
    display:table;
}


.avatar-module {
    display: table-cell;
    vertical-align: middle;
}


.avatar-module img {
    border-radius: 50%!important;
    height: 120px;
    width: 120px;
    float: left;
    margin:0px 20px 0px 20px;
}


.article-footer p {
    line-height:1.5em;
    padding-left:15px;
}


@media (max-width:750px) {
    .article-footer {
        display:inherit;
        margin:0px;
    }

    .avatar-module {
        display:inherit;
        vertical-align: none;
    }

    .avatar-module img {
        margin-left: auto;
        margin-right: auto;
        display: block;
        float:none;
        margin-top:0px;
    }

    .article-footer p {
        text-align:center;
        padding-left:0px;
    }
}

    </style>
<style>

 h1 {
     font-weight: bold;
     font-size: 25px;
 }

</style>
</head>
<body>
    <a href="#content" class="sr-only sr-only-focusable">Skip to main content</a>

    <!-- Menubar -->
    <nav class="navbar navbar-expand-md static-top mb-4
                navbar-dark
                bg-dark
                "><div class="container">
<!-- This keeps the margins nice -->
            <a class="navbar-brand" href="../../">

                    <span id="blog-title">Avil Page</span>
            </a>
            <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#bs-navbar" aria-controls="bs-navbar" aria-expanded="false" aria-label="Toggle navigation">
                <span class="navbar-toggler-icon"></span>
            </button>

            <div class="collapse navbar-collapse" id="bs-navbar">
                <ul class="navbar-nav mr-auto">
<li class="nav-item">
<a href="../../archive.html" class="nav-link">Archive</a>
                </li>
<li class="nav-item">
<a href="../../top-10.html" class="nav-link">Top 10</a>
                </li>
<li class="nav-item">
<a href="../../rss.xml" class="nav-link">RSS</a>

                    
                </li>
</ul>
<ul class="navbar-nav navbar-right"></ul>
</div>
<!-- /.navbar-collapse -->
        </div>
<!-- /.container -->
    </nav><!-- End of Menubar --><div class="container" id="content" role="main">
    <div class="row justify-content-md-center">
        <div class="body-content col col-lg-6">
            <!--Body content-->
            
    <article class="post-text h-entry hentry postpage" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title" itemprop="headline name"><a href="#" class="u-url">Mastering Kraken2 - Part 1 - Initial Runs</a></h1>

        <div class="metadata">
            <p class="byline author vcard p-author h-card">
                <span class="byline-name fn p-name" itemprop="author">
                    Chillar Anand
                </span>
            </p>

            <p class="dateline">
                    8 min read
            </p>

            <p class="dateline">
                <a href="#" rel="bookmark">
                    <time class="published dt-published" datetime="2024-07-28T10:44:25+05:30" itemprop="datePublished" title="2024-07-28">2024-07-28</time></a>
            </p>
            
        </div>

        
    </header><div class="e-content entry-content" itemprop="articleBody text">
            <h4>Mastering Kraken2</h4>
<p>Part 1 - Initial Runs (this post)</p>
<p><a href="mastering-kraken2-performance-optimisation.html">Part 2 - Performance Optimisation</a></p>
<p>Part 3 - Custom Indices (upcoming)</p>
<h4>Introduction</h4>
<p>Kraken2<sup id="fnref:Kraken2"><a class="footnote-ref" href="mastering-kraken2-initial-runs.html#fn:Kraken2">1</a></sup> is widely used for metagenomics taxonomic classification, and it has pre-built indexes for many organisms. In this series, we will learn</p>
<ul>
<li>How to set up kraken2, download pre-built indices</li>
<li>Run kraken2 (8GB RAM) at ~0.19 Mbp/m (million base pairs per minute)</li>
<li>Learn various ways to speed up the classification process</li>
<li>Run kraken2 (128GB RAM) at ~1200 Mbp/m</li>
<li>Build custom indices</li>
</ul>
<h4>Installation</h4>
<p>Let's start with an 8 GB RAM machine. We can install kraken2 using the <code>install_kraken2.sh</code> script as per the manual<sup id="fnref:install_kraken2"><a class="footnote-ref" href="mastering-kraken2-initial-runs.html#fn:install_kraken2">2</a></sup>.</p>
<pre class="code literal-block">$ git clone https://github.com/DerrickWood/kraken2
$ <span class="nb">cd</span> kraken2
$ ./install_kraken2.sh /usr/local/bin
<span class="c1"># ensure kraken2 is in the PATH</span>
$ <span class="nb">export</span> <span class="nv">PATH</span><span class="o">=</span><span class="nv">$PATH</span>:/usr/local/bin
</pre>
<p>If you already have conda installed, you can install kraken2 from conda as well.</p>
<pre class="code literal-block">$ conda install -c bioconda kraken2
</pre>
<h4>Download pre-built indices</h4>
<p>Building kraken2 indices take a lot of time and resources. For now, let's download and use the pre-built indices. In the final post, we will learn how to build the indices.</p>
<p>Genomic Index Zone<sup id="fnref:GenomicIndexZone"><a class="footnote-ref" href="mastering-kraken2-initial-runs.html#fn:GenomicIndexZone">3</a></sup> provides pre-built indices for kraken2. Let's download the standard database. It contains Refeq archaea, bacteria, viral, plasmid, human1, &amp; UniVec_Core. </p>
<pre class="code literal-block">$ wget https://genome-idx.s3.amazonaws.com/kraken/k2_standard_20240605.tar.gz
$ mkdir k2_standard
$ tar -xvf k2_standard_20240605.tar.gz -C k2_standard
</pre>
<p>The extracted directory contains three files - <code>hash.k2d</code>, <code>opts.k2d</code>, <code>taxo.k2d</code> which are the kraken2 database files.</p>
<pre class="code literal-block">$ ls -l *.k2d
.rw-r--r--  83G anand <span class="m">13</span> Jul <span class="m">12</span>:34 hash.k2d
.rw-r--r--   <span class="m">64</span> anand <span class="m">13</span> Jul <span class="m">12</span>:34 opts.k2d
.rw-r--r-- <span class="m">4</span>.0M anand <span class="m">13</span> Jul <span class="m">12</span>:34 taxo.k2d
</pre>
<h4>Classification</h4>
<p>To run the taxonomic classification, let's use <code>ERR10359977</code> human gut meta genome from NCBI SRA.</p>
<pre class="code literal-block">$ wget https://ftp.sra.ebi.ac.uk/vol1/fastq/ERR103/077/ERR10359977/ERR10359977.fastq.gz
$ kraken2 --db k2_standard --report report.txt ERR10359977.fastq.gz &gt; output.txt
</pre>
<p>By default, the machine I have used has 8GB RAM and an additioinal 8GB swap. Since kraken2 needs entire db(~80GB) in memory, when the process tries to consume more than 16GB memory, the kernel will kill the process. </p>
<pre class="code literal-block">$ <span class="nb">time</span> kraken2 --db k2_standard --paired SRR6915097_1.fastq.gz SRR6915097_2.fastq.gz &gt; output.txt
Loading database information...Command terminated by signal <span class="m">9</span>
<span class="m">0</span>.02user <span class="m">275</span>.83system <span class="m">8</span>:17.43elapsed <span class="m">55</span>%CPU 
</pre>
<p>To prevent this, let's increase the swap space to 128 GB.</p>
<pre class="code literal-block"><span class="c1"># Create an empty swapfile of 128GB</span>
sudo dd <span class="k">if</span><span class="o">=</span>/dev/zero <span class="nv">of</span><span class="o">=</span>/swapfile <span class="nv">bs</span><span class="o">=</span>1G <span class="nv">count</span><span class="o">=</span><span class="m">128</span>

<span class="c1"># Turn swap off - It might take several minutes</span>
sudo swapoff -a

<span class="c1"># Set the permissions for swapfile</span>
sudo chmod <span class="m">0600</span> /swapfile

<span class="c1"># make it a swap area</span>
sudo mkswap /swapfile  

<span class="c1"># Turn the swap on</span>
sudo swapon /swapfile
</pre>
<p>We can time the classification process using the <code>time</code> command.</p>
<pre class="code literal-block">$ <span class="nb">time</span> kraken2 --db k2_standard --report report.txt ERR10359977.fastq.gz &gt; output.txt
</pre>
<p>If you have a machine with large RAM, the same scenario can be simulated using <code>systemd-run</code>. This will limit the memory usage of kraken2 to 6.5GB. </p>
<pre class="code literal-block">$ <span class="nb">time</span> systemd-run --scope -p <span class="nv">MemoryMax</span><span class="o">=</span><span class="m">6</span>.5G --user <span class="nb">time</span> kraken2 --db k2_standard --report report.txt ERR10359977.fastq.gz &gt; output.txt
</pre>
<p>Depending on the CPU performance, this will take around ~40 minutes to complete.</p>
<pre class="code literal-block">Loading database information... <span class="k">done</span>.
<span class="m">95064</span> sequences <span class="o">(</span><span class="m">14</span>.35 Mbp<span class="o">)</span> processed <span class="k">in</span> <span class="m">1026</span>.994s <span class="o">(</span><span class="m">5</span>.6 Kseq/m, <span class="m">0</span>.84 Mbp/m<span class="o">)</span>.
  <span class="m">94939</span> sequences classified <span class="o">(</span><span class="m">99</span>.87%<span class="o">)</span>
  <span class="m">125</span> sequences unclassified <span class="o">(</span><span class="m">0</span>.13%<span class="o">)</span>
  <span class="m">4</span>.24user <span class="m">658</span>.68system <span class="m">38</span>:26.78elapsed <span class="m">28</span>%CPU 
</pre>
<p>If we try gut WGS(Whole Genome Sequence) sample like <code>SRR6915097</code> <sup id="fnref:srr1"><a class="footnote-ref" href="mastering-kraken2-initial-runs.html#fn:srr1">4</a></sup><sup id="fnref:srr2"><a class="footnote-ref" href="mastering-kraken2-initial-runs.html#fn:srr2">5</a></sup>. which contains ~3.3 Gbp, it will take weeks to complete.</p>
<pre class="code literal-block">$ wget -c https://ftp.sra.ebi.ac.uk/vol1/fastq/SRR691/007/SRR6915097/SRR6915097_1.fastq.gz
$ wget -c https://ftp.sra.ebi.ac.uk/vol1/fastq/SRR691/007/SRR6915097/SRR6915097_2.fastq.gz

$ <span class="nb">time</span> systemd-run --scope -p <span class="nv">MemoryMax</span><span class="o">=</span>6G --user <span class="nb">time</span> kraken2 --db k2_standard --paired SRR6915097_1.fastq.gz SRR6915097_2.fastq.gz &gt; output.txt
</pre>
<p>I tried running this on 8 GB machine. Even after 10 days, it processed only 10% of the data.</p>
<p>If we have to process a large number of such samples, it takes months and this is not a practical solution. </p>
<h4>Conclusion</h4>
<p>In this post, we ran kraken2 on an 8GB machine and learned that it is not feasible to run kraken2 on large samples.</p>
<p>In the next post, we will learn how to speed up the classification process and run classification at 1200 Mbp/m.</p>
<p><strong>Next</strong>: <a href="mastering-kraken2-performance-optimisation.html">Part 2 - Performance Optimisation</a></p>
<div class="footnote">
<hr>
<ol>
<li id="fn:Kraken2">
<p><a href="https://ccb.jhu.edu/software/kraken2/">Kraken2</a> <a class="footnote-backref" href="mastering-kraken2-initial-runs.html#fnref:Kraken2" title="Jump back to footnote 1 in the text">↩</a></p>
</li>
<li id="fn:install_kraken2">
<p><a href="https://github.com/DerrickWood/kraken2/blob/master/docs/MANUAL.markdown#installation">Kraken2 - Manual - Install</a> <a class="footnote-backref" href="mastering-kraken2-initial-runs.html#fnref:install_kraken2" title="Jump back to footnote 2 in the text">↩</a></p>
</li>
<li id="fn:GenomicIndexZone">
<p><a href="https://benlangmead.github.io/aws-indexes/k2">Genomic Index Zone - k2</a> <a class="footnote-backref" href="mastering-kraken2-initial-runs.html#fnref:GenomicIndexZone" title="Jump back to footnote 3 in the text">↩</a></p>
</li>
<li id="fn:srr1">
<p><a href="https://ftp.sra.ebi.ac.uk/vol1/fastq/SRR691/007/SRR6915097/SRR6915097_1.fastq.gz">SRR6915097_1.fastq.gz</a> <a class="footnote-backref" href="mastering-kraken2-initial-runs.html#fnref:srr1" title="Jump back to footnote 4 in the text">↩</a></p>
</li>
<li id="fn:srr2">
<p><a href="https://ftp.sra.ebi.ac.uk/vol1/fastq/SRR691/007/SRR6915097/SRR6915097_2.fastq.gz">SRR6915097_1.fastq.gz</a> <a class="footnote-backref" href="mastering-kraken2-initial-runs.html#fnref:srr2" title="Jump back to footnote 5 in the text">↩</a></p>
</li>
</ol>
</div>
        </div>
        <aside class="postpromonav"><nav><ul itemprop="keywords" class="tags">
<li><a class="tag p-category" href="../../tags/devops.html" rel="tag">devops</a></li>
            <li><a class="tag p-category" href="../../tags/kraken2.html" rel="tag">kraken2</a></li>
            <li><a class="tag p-category" href="../../tags/metagenomics.html" rel="tag">metagenomics</a></li>
        </ul>
<ul class="pager hidden-print">
<li class="previous">
                <a href="../06/headlamp-k8s-lens-open-source-alternative.html" rel="prev" title="Headlamp - k8s Lens open source alternative">Previous post</a>
            </li>
            <li class="next">
                <a href="mastering-kraken2-performance-optimisation.html" rel="next" title="Mastering kraken2 - Part 2 - Performance Optimisation">Next post</a>
            </li>
        </ul></nav></aside><div class="full-article-footer">
            <div class="article-footer">

                <div class="avatar-module">
                    <img class="avatar" height="100px" src="../../images/chillaranand.jpg">
</div>

                <p class="avatar-module">
                    <b>Chillar Anand</b>
                    <br>
                    A blog about python, careers &amp; life.
                    <br>
                    To contact me, <a href="https://forms.gle/Hre4z4aLqJA5zYWe6">send a message here</a>.
                </p>

            </div>
        </div>

        
    </article><!--End of body content--><footer id="footer"><div class="container align-items-center justify-content-center d-flex">

<footer class="footer"><a href="https://github.com/ChillarAnand">
<img src="../../images/icons8-github.svg" alt="github-chillar-anand" height="34"></a>

  

<a href="https://stackoverflow.com/users/2698552/chillar-anand">
<img src="../../images/icons8-so.svg" alt="github-chillar-anand" height="30"></a>

  

<a href="https://youtube.com/@avilpage">
<img src="../../images/icons8-youtube.svg" alt="github-chillar-anand" height="34"></a>

  

<a href="https://linkedin.com/in/chillaranand">
<img src="../../images/icons8-linkedin.svg" alt="github-chillar-anand" height="34"></a>

  

<a href="https://twitter.com/chillaranand">
<img src="../../images/icons8-twitter.svg" alt="github-chillar-anand" height="34"></a>

</footer>
</div>

<br><br></footer>
</div>
    </div>
    </div>

                <script src="../../assets/js/jquery.min.js"></script><script src="../../assets/js/popper.min.js"></script><script src="../../assets/js/bootstrap.min.js"></script><script src="../../assets/js/baguetteBox.min.js"></script><script>
    baguetteBox.run('div#content', {
        ignoreClass: 'islink',
        captions: function(element){var i=element.getElementsByTagName('img')[0];return i===undefined?'':i.alt;}});
    </script>
</body>
</html>
