<!DOCTYPE html>
<html prefix="
og: http://ogp.me/ns# article: http://ogp.me/ns/article#
" lang="en">
<head>
<meta charset="utf-8">
<meta name="description" content="Avil Page - Personal &amp; tech blog by Chillar Anand">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Avil Page</title>
<link href="../assets/css/bootstrap.min.css" rel="stylesheet" type="text/css">
<link href="../assets/css/baguetteBox.min.css" rel="stylesheet" type="text/css">
<link href="../assets/css/rst.css" rel="stylesheet" type="text/css">
<link href="../assets/css/code.css" rel="stylesheet" type="text/css">
<link href="../assets/css/theme.css" rel="stylesheet" type="text/css">
<meta name="theme-color" content="#5670d4">
<meta name="generator" content="Nikola (getnikola.com)">
<link rel="alternate" type="application/rss+xml" title="RSS" hreflang="en" href="../rss.xml">
<link rel="canonical" href="https://avilpage.com/blog/">
<link rel="next" href="index-19.html" type="text/html">
<!--[if lt IE 9]><script src="../assets/js/html5.js"></script><![endif]--><style>

 h1 {
     font-weight: bold;
     font-size: 25px;
 }

</style>
</head>
<body>
    <a href="#content" class="sr-only sr-only-focusable">Skip to main content</a>

    <!-- Menubar -->
    <nav class="navbar navbar-expand-md static-top mb-4
                navbar-dark
                bg-dark
                "><div class="container">
<!-- This keeps the margins nice -->
            <a class="navbar-brand" href="../">

                    <span id="blog-title">Avil Page</span>
            </a>
            <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#bs-navbar" aria-controls="bs-navbar" aria-expanded="false" aria-label="Toggle navigation">
                <span class="navbar-toggler-icon"></span>
            </button>

            <div class="collapse navbar-collapse" id="bs-navbar">
                <ul class="navbar-nav mr-auto">
<li class="nav-item">
<a href="../archive.html" class="nav-link">Archive</a>
                </li>
<li class="nav-item">
<a href="../top-10.html" class="nav-link">Top 10</a>
                </li>
<li class="nav-item">
<a href="../rss.xml" class="nav-link">RSS</a>

                    
                </li>
</ul>
<ul class="navbar-nav navbar-right"></ul>
</div>
<!-- /.navbar-collapse -->
        </div>
<!-- /.container -->
    </nav><!-- End of Menubar --><div class="container" id="content" role="main">
    <div class="row justify-content-md-center">
        <div class="body-content col col-lg-6">
            <!--Body content-->
            
    
    
<div class="postindex">
    <article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="../2024/07/mastering-kraken2-build-custom-db.html" class="u-url">Mastering Kraken2 - Part 3 - Build Custom Database</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                Chillar Anand
            </span></p>
            <p class="dateline">
            <a href="../2024/07/mastering-kraken2-build-custom-db.html" rel="bookmark">
            <time class="published dt-published" datetime="2024-08-01T10:52:30+05:30" itemprop="datePublished" title="2024-08-01">2024-08-01</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <h4>Mastering Kraken2</h4>
<p><a href="../2024/07/mastering-kraken2-initial-runs.html">Part 1 - Initial Runs</a></p>
<p><a href="../2024/07/mastering-kraken2-performance-optimisation.html">Part 2 - Performance Optimisation</a></p>
<p><a href="../2024/07/mastering-kraken2-build-custom-db.html">Part 3 - Building custom databases</a> (this post)</p>
<p>Part 4 - Regular vs Fast Builds (upcoming)</p>
<p>Part 5 - Benchmarking (upcoming)</p>
<h4>Introduction</h4>
<p>In the previous post, we learned how to improve kraken2<sup id="fnref:k2"><a class="footnote-ref" href="../2024/07/mastering-kraken2-build-custom-db.html#fn:k2">1</a></sup> classification performance. So far we have downloaded &amp; used pre-built genome indices(databases). </p>
<p>In this post, let's build a custom database for kraken2. For simplicity, let's use only refseq archaea genomes<sup id="fnref:rag"><a class="footnote-ref" href="../2024/07/mastering-kraken2-build-custom-db.html#fn:rag">2</a></sup> for building the index.</p>
<h4>Building Custom Database</h4>
<p>First, we need to download the taxonomy files. We can use the <code>k2</code> script provided by kraken2.</p>
<pre class="code literal-block">$ k2 download-taxonomy --db custom_db
</pre>
<p>This takes ~30 minutes depending on the network speed. The taxonomy files are downloaded to the <code>custom_db/taxonomy</code> directory.</p>
<pre class="code literal-block">$ ls custom_db/taxonomy
citations.dmp  division.dmp  gencode.dmp  merged.dmp  nodes.dmp
nucl_wgs.accession2taxid delnodes.dmp  gc.prt 
images.dmp  names.dmp  nucl_gb.accession2taxid  readme.txt

$ du -hs custom_db/taxonomy
43G     custom_db/taxonomy
</pre>
<p>For simplicity, let's use the archaea refseq genomes. We can use <code>kraken2-build</code> to download the refseq genomes.</p>
<pre class="code literal-block">$ k2 download-library --library archaea --db custom_db
</pre>
<p>This runs on a single thread. Instead of using <code>kraken2-build</code>, we can use <code>ncbi-genome-download</code><sup id="fnref:ngd"><a class="footnote-ref" href="../2024/07/mastering-kraken2-build-custom-db.html#fn:ngd">3</a></sup> tool to download the genomes. This provides much granular control over the download process. For example, we can download only <code>--assembly-levels complete</code> genomes. We can also download multiple genomes in parallel.</p>
<pre class="code literal-block">$ pip install ncbi-genome-download

$ conda install -c bioconda ncbi-genome-download

$ ncbi-genome-download -s refseq -F fasta --parallel <span class="m">40</span> -P archaea
Checking assemblies: <span class="m">100</span>%<span class="p">|</span>███<span class="p">|</span> <span class="m">2184</span>/2184 <span class="o">[</span><span class="m">00</span>:19&lt;<span class="m">00</span>:00, <span class="m">111</span>.60entries/s<span class="o">]</span>
Downloading assemblies: <span class="m">100</span>%<span class="p">|</span>███<span class="p">|</span> <span class="m">2184</span>/2184  <span class="o">[</span><span class="m">02</span>:04&lt;<span class="m">00</span>:00,  <span class="m">4</span>.54s/files<span class="o">]</span>
Downloading assemblies: 2184files <span class="o">[</span><span class="m">02</span>:23, 2184files/s<span class="o">]</span>
</pre>
<p>In just 2 minutes, it has downloaded all the files. Lets gunzip the files.</p>
<pre class="code literal-block">$ find refseq -name <span class="s2">"*.gz"</span> -print0 <span class="p">|</span> parallel -0 gunzip

$ du -hs refseq
<span class="m">5</span>.9G    refseq
</pre>
<p>Lets add all fasta genome files to the custom database</p>
<pre class="code literal-block">$ <span class="nb">time</span> find refseq -name <span class="s2">"*.fna"</span> -exec kraken2-build --add-to-library <span class="o">{}</span> --db custom_db <span class="se">\;</span>
<span class="m">667</span>.46s user <span class="m">90</span>.78s system <span class="m">106</span>% cpu <span class="m">12</span>:54.80 total
</pre>
<p><code>kraken2-build</code> doesn't use multiple threads for adding genomes to the database. In addition to that, it also doesn't check if the genome is already present in the database. </p>
<p>Let's use <code>k2</code> for adding genomes to the database.</p>
<pre class="code literal-block"><span class="nb">export</span> <span class="nv">KRAKEN_NUM_THREADS</span><span class="o">=</span><span class="m">40</span>

$ find . -name <span class="s2">"*.fna"</span> -exec k2 add-to-library --files <span class="o">{}</span> --db custom_db <span class="se">\;</span>
<span class="m">668</span>.37s user <span class="m">88</span>.44s system <span class="m">159</span>% cpu <span class="m">7</span>:54.40 total
</pre>
<p>This took only half the time compared to <code>kraken2-build</code>.</p>
<p>Let's build the index from the library.</p>
<pre class="code literal-block">$ <span class="nb">time</span> kraken2-build --db custom_db --build --threads <span class="m">36</span>
Creating sequence ID to taxonomy ID map <span class="o">(</span>step <span class="m">1</span><span class="o">)</span>...
Found <span class="m">0</span>/125783 targets, searched through <span class="m">60000000</span> accession IDs...
Found <span class="m">59923</span>/125783 targets, searched through <span class="m">822105735</span> accession IDs, search complete.
lookup_accession_numbers: <span class="m">65860</span>/125783 accession numbers remain unmapped, see unmapped.txt <span class="k">in</span> DB directory
Sequence ID to taxonomy ID map complete. <span class="o">[</span>2m1.950s<span class="o">]</span>
Estimating required capacity <span class="o">(</span>step <span class="m">2</span><span class="o">)</span>...
Estimated <span class="nb">hash</span> table requirement: <span class="m">5340021028</span> bytes
Capacity estimation complete. <span class="o">[</span><span class="m">23</span>.875s<span class="o">]</span>
Building database files <span class="o">(</span>step <span class="m">3</span><span class="o">)</span>...
Taxonomy parsed and converted.
CHT created with <span class="m">11</span> bits reserved <span class="k">for</span> taxid.
Completed processing of <span class="m">59911</span> sequences, <span class="m">3572145823</span> bp
Writing data to disk...  complete.
Database files completed. <span class="o">[</span>12m3.368s<span class="o">]</span>
Database construction complete. <span class="o">[</span>Total: 14m29.666s<span class="o">]</span>
kraken2-build --db custom_db --build --threads <span class="m">36</span>  <span class="m">24534</span>.98s user <span class="m">90</span>.50s system <span class="m">2831</span>% cpu <span class="m">14</span>:29.75 total

$ ls -ll
.rw-rw-r-- <span class="m">5</span>.3G anand  <span class="m">1</span> Aug <span class="m">16</span>:35 hash.k2d
drwxrwxr-x    - anand  <span class="m">1</span> Aug <span class="m">12</span>:32 library
.rw-rw-r--   <span class="m">64</span> anand  <span class="m">1</span> Aug <span class="m">16</span>:35 opts.k2d
.rw-rw-r-- <span class="m">1</span>.5M anand  <span class="m">1</span> Aug <span class="m">16</span>:22 seqid2taxid.map
.rw-rw-r-- 115k anand  <span class="m">1</span> Aug <span class="m">16</span>:23 taxo.k2d
lrwxrwxrwx   <span class="m">20</span> anand  <span class="m">1</span> Aug <span class="m">12</span>:31 taxonomy
.rw-rw-r-- <span class="m">1</span>.2M anand  <span class="m">1</span> Aug <span class="m">16</span>:22 unmapped.txt
</pre>
<p>We are able to build index for ~6GB input files in ~15 minutes.</p>
<h4>Conclusion</h4>
<p>We learnt some useful tips to speed up the custom database creation process. In the next post, we will learn about regular vs. fast builds.</p>
<div class="footnote">
<hr>
<ol>
<li id="fn:k2">
<p><a href="https://ccb.jhu.edu/software/kraken2/">Kraken2</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-build-custom-db.html#fnref:k2" title="Jump back to footnote 1 in the text">↩</a></p>
</li>
<li id="fn:rag">
<p><a href="https://ftp.ncbi.nlm.nih.gov/genomes/refseq/archaea/">RefSeq Archaea genomes</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-build-custom-db.html#fnref:rag" title="Jump back to footnote 2 in the text">↩</a></p>
</li>
<li id="fn:ngd">
<p><a href="https://github.com/kblin/ncbi-genome-download">https://github.com/kblin/ncbi-genome-download</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-build-custom-db.html#fnref:ngd" title="Jump back to footnote 3 in the text">↩</a></p>
</li>
</ol>
</div>
    </div>
    </article><article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="../2024/07/mastering-kraken2-performance-optimisation.html" class="u-url">Mastering Kraken2 - Part 2 - Performance Optimisation</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                Chillar Anand
            </span></p>
            <p class="dateline">
            <a href="../2024/07/mastering-kraken2-performance-optimisation.html" rel="bookmark">
            <time class="published dt-published" datetime="2024-07-28T10:51:30+05:30" itemprop="datePublished" title="2024-07-28">2024-07-28</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <h4>Mastering Kraken2</h4>
<p><a href="../2024/07/mastering-kraken2-initial-runs.html">Part 1 - Initial Runs</a></p>
<p><a href="../2024/07/mastering-kraken2-performance-optimisation.html">Part 2 - Performance Optimisation</a> (this post)</p>
<p><a href="../2024/07/mastering-kraken2-build-custom-db.html">Part 3 - Building custom databases</a></p>
<p>Part 4 - Regular vs Fast Builds (upcoming)</p>
<p>Part 5 - Benchmarking (upcoming)</p>
<h4>Introduction</h4>
<p>In the previous post, we learned how to set up kraken2<sup id="fnref:k2"><a class="footnote-ref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fn:k2">1</a></sup>, download pre-built indices, and run kraken2. In this post, we will learn various ways to speed up the classification process.</p>
<h4>Increasing RAM</h4>
<p>Kraken2 standard database is ~80GB in size. It is recommended to have at least db size RAM to run kraken2 efficiently<sup id="fnref:ksr"><a class="footnote-ref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fn:ksr">2</a></sup>. Let's use 128GB RAM machine and run kraken2 with ERR10359977<sup id="fnref:err"><a class="footnote-ref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fn:err">3</a></sup> sample.</p>
<pre class="code literal-block">$ <span class="nb">time</span> kraken2 --db k2_standard --report report.txt ERR10359977.fastq.gz &gt; output.txt
Loading database information... <span class="k">done</span>.
<span class="m">95064</span> sequences <span class="o">(</span><span class="m">14</span>.35 Mbp<span class="o">)</span> processed <span class="k">in</span> <span class="m">2</span>.142s <span class="o">(</span><span class="m">2662</span>.9 Kseq/m, <span class="m">402</span>.02 Mbp/m<span class="o">)</span>.
  <span class="m">94816</span> sequences classified <span class="o">(</span><span class="m">99</span>.74%<span class="o">)</span>
  <span class="m">248</span> sequences unclassified <span class="o">(</span><span class="m">0</span>.26%<span class="o">)</span>
kraken2 --db k2_standard --report report.txt ERR10359977.fastq.gz &gt;   <span class="m">1</span>.68s user <span class="m">152</span>.19s system <span class="m">35</span>% cpu <span class="m">7</span>:17.55 total
</pre>
<p>Now the time taken has come down from 40 minutes to 7 minutes. The classification speed has also increased from 0.19 Mbp/m to 402.02 Mbp/m.</p>
<p>The previous sample had only a few reads, and the speed is not a good indicator. Let's run kraken2 with a larger sample.</p>
<pre class="code literal-block">$ <span class="nb">time</span> kraken2 --db k2_standard --report report.txt --paired SRR6915097_1.fastq.gz SRR6915097_2.fastq.gz &gt; output.txt
Loading database information... <span class="k">done</span>.
Processed <span class="m">14980000</span> sequences <span class="o">(</span><span class="m">2972330207</span> bp<span class="o">)</span> ...
<span class="m">17121245</span> sequences <span class="o">(</span><span class="m">3397</span>.15 Mbp<span class="o">)</span> processed <span class="k">in</span> <span class="m">797</span>.424s <span class="o">(</span><span class="m">1288</span>.2 Kseq/m, <span class="m">255</span>.61 Mbp/m<span class="o">)</span>.
  <span class="m">9826671</span> sequences classified <span class="o">(</span><span class="m">57</span>.39%<span class="o">)</span>
  <span class="m">7294574</span> sequences unclassified <span class="o">(</span><span class="m">42</span>.61%<span class="o">)</span>
kraken2 --db k2_standard --report report.txt --paired &gt; output.txt  <span class="m">526</span>.39s user <span class="m">308</span>.24s system <span class="m">68</span>% cpu <span class="m">20</span>:23.86 total
</pre>
<p>This took almost 20 minutes to classify ~3 Gbp of data. Out of 20 minutes, 13 minutes was spent in classification. The remaining time in loading the db into memory.</p>
<p>Let's use k2_plusPF<sup id="fnref:k2p"><a class="footnote-ref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fn:k2p">4</a></sup> db, which is twice the size of k2_standard and run kraken2.</p>
<pre class="code literal-block">$ <span class="nb">time</span> kraken2 --db k2_plusfp --report report.txt --paired SRR6915097_1.fastq.gz SRR6915097_2.fastq.gz &gt; output.txt
Loading database information...done.
<span class="m">17121245</span> sequences <span class="o">(</span><span class="m">3397</span>.15 Mbp<span class="o">)</span> processed <span class="k">in</span> <span class="m">755</span>.290s <span class="o">(</span><span class="m">1360</span>.1 Kseq/m, <span class="m">269</span>.87 Mbp/m<span class="o">)</span>.
  <span class="m">9903824</span> sequences classified <span class="o">(</span><span class="m">57</span>.85%<span class="o">)</span>
  <span class="m">7217421</span> sequences unclassified <span class="o">(</span><span class="m">42</span>.15%<span class="o">)</span>
kraken2 --db k2_plusfp/ --report report.txt --paired SRR6915097_1.fastq.gz  &gt;   <span class="m">509</span>.71s user <span class="m">509</span>.51s system <span class="m">55</span>% cpu <span class="m">30</span>:35.49 total
</pre>
<p>This took ~30 minutes to complete, but the classification took only 13 minutes similar to k2_standard. The remaining time was spent in loading the db into memory.</p>
<h4>Preloading db into RAM</h4>
<p>We can use vmtouch<sup id="fnref:vmt"><a class="footnote-ref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fn:vmt">5</a></sup> to preload db into RAM. kraken2 provides <code>--memory-mapping</code> option to use preloaded db. </p>
<pre class="code literal-block">$ vmtouch -vt k2_standard/hash.k2d k2_standard/opts.k2d k2_standard/taxo.k2d
           Files: <span class="m">3</span>
     Directories: <span class="m">0</span>
   Touched Pages: <span class="m">20382075</span> <span class="o">(</span>77G<span class="o">)</span>
         Elapsed: <span class="m">434</span>.77 seconds
</pre>
<p>Now, let's run kraken2 with <code>--memory-mapping</code> option.</p>
<pre class="code literal-block">$ <span class="nb">time</span> kraken2 --db k2_standard --report report.txt --memory-mapping --paired SRR6915097_1.fastq.gz SRR6915097_2.fastq.gz &gt; output.txt
Loading database information... <span class="k">done</span>.
<span class="m">17121245</span> sequences <span class="o">(</span><span class="m">3397</span>.15 Mbp<span class="o">)</span> processed <span class="k">in</span> <span class="m">532</span>.486s <span class="o">(</span><span class="m">1929</span>.2 Kseq/m, <span class="m">382</span>.79 Mbp/m<span class="o">)</span>.
  <span class="m">9826671</span> sequences classified <span class="o">(</span><span class="m">57</span>.39%<span class="o">)</span>
  <span class="m">7294574</span> sequences unclassified <span class="o">(</span><span class="m">42</span>.61%<span class="o">)</span>
  kraken2 --db k2_standard --report report.txt --paired SRR6915097_1.fastq.gz   &gt;  <span class="m">424</span>.20s user <span class="m">11</span>.76s system <span class="m">81</span>% cpu <span class="m">8</span>:54.98 total
</pre>
<p>Now the classification took only ~10 minutes.</p>
<h4>Multi threading</h4>
<p>kraken2 supports multiple threads. I am using a machine with 40 threads.</p>
<pre class="code literal-block">$ <span class="nb">time</span> kraken2 --db k2_standard --report report.txt --paired SRR6915097_1.fastq.gz SRR6915097_2.fastq.gz --memory-mapping --threads <span class="m">32</span> &gt; output.txt
Loading database information... <span class="k">done</span>.
<span class="m">17121245</span> sequences <span class="o">(</span><span class="m">3397</span>.15 Mbp<span class="o">)</span> processed <span class="k">in</span> <span class="m">71</span>.675s <span class="o">(</span><span class="m">14332</span>.5 Kseq/m, <span class="m">2843</span>.81 Mbp/m<span class="o">)</span>.
  <span class="m">9826671</span> sequences classified <span class="o">(</span><span class="m">57</span>.39%<span class="o">)</span>
  <span class="m">7294574</span> sequences unclassified <span class="o">(</span><span class="m">42</span>.61%<span class="o">)</span>
kraken2 --db k2_standard --report report.txt --paired SRR6915097_1.fastq.gz      <span class="m">556</span>.58s user <span class="m">22</span>.85s system <span class="m">762</span>% cpu <span class="m">1</span>:16.02 total
</pre>
<p>With 32 threads, the classification took only 1 minute. Beyond 32 threads, the classification time did not decrease significantly.</p>
<h4>Optimising input files</h4>
<p>So far we have used gzipped input files. Let's use unzipped input files and run kraken2.</p>
<pre class="code literal-block">$ gunzip SRR6915097_1.fastq.gz
$ gunzip SRR6915097_2.fastq.gz

$ <span class="nb">time</span> kraken2 --db k2_standard --report report.txt --paired SRR6915097_1.fastq SRR6915097_2.fastq --memory-mapping --threads <span class="m">30</span> &gt; output.txt
Loading database information... <span class="k">done</span>.
<span class="m">17121245</span> sequences <span class="o">(</span><span class="m">3397</span>.15 Mbp<span class="o">)</span> processed <span class="k">in</span> <span class="m">34</span>.809s <span class="o">(</span><span class="m">29512</span>.0 Kseq/m, <span class="m">5855</span>.68 Mbp/m<span class="o">)</span>.
  <span class="m">9826671</span> sequences classified <span class="o">(</span><span class="m">57</span>.39%<span class="o">)</span>
  <span class="m">7294574</span> sequences unclassified <span class="o">(</span><span class="m">42</span>.61%<span class="o">)</span>
kraken2 --db k2_standard --report report.txt --paired SRR6915097_1.fastq    <span class="m">30</span>   <span class="m">565</span>.03s user <span class="m">17</span>.12s system <span class="m">1530</span>% cpu <span class="m">38</span>.047 total
</pre>
<p>Now the classification time has come down to 40 seconds.</p>
<p>Since the input fastq files are paired, interleaving the files also takes time. Lets interleave the files and run kraken2.</p>
<p>To interleave the files, lets use <code>seqfu</code> tool.</p>
<pre class="code literal-block">$ conda install -y -c conda-forge -c bioconda <span class="s2">"seqfu&gt;1.10"</span>

$ seqfu interleave -1 SRR6915097_1.fastq.gz -2 SRR6915097_2.fastq.gz &gt; SRR6915097.fastq

$ <span class="nb">time</span> kraken2 --db k2_standard --report report.txt --memory-mapping SRR6915097.fq --threads <span class="m">32</span> &gt; output.txt
Loading database information... <span class="k">done</span>.
<span class="m">34242490</span> sequences <span class="o">(</span><span class="m">3397</span>.15 Mbp<span class="o">)</span> processed <span class="k">in</span> <span class="m">20</span>.199s <span class="o">(</span><span class="m">101714</span>.1 Kseq/m, <span class="m">10090</span>.91 Mbp/m<span class="o">)</span>.
  <span class="m">17983321</span> sequences classified <span class="o">(</span><span class="m">52</span>.52%<span class="o">)</span>
  <span class="m">16259169</span> sequences unclassified <span class="o">(</span><span class="m">47</span>.48%<span class="o">)</span>
kraken2 --db k2_standard --report report.txt --memory-mapping SRR6915097.fq  <span class="m">32</span>  <span class="m">618</span>.96s user <span class="m">18</span>.24s system <span class="m">2653</span>% cpu <span class="m">24</span>.013 total
</pre>
<p>Now the classification time has come down to 24 seconds. </p>
<h4>Conclusion</h4>
<p>In terms of classification speed, we have come a long way from 0.1 Mbp/m to 1200 Mbp/m. In the next post, we will learn how to optimise the creation of custom indices.</p>
<div class="footnote">
<hr>
<ol>
<li id="fn:k2">
<p><a href="https://ccb.jhu.edu/software/kraken2/">Kraken2</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fnref:k2" title="Jump back to footnote 1 in the text">↩</a></p>
</li>
<li id="fn:ksr">
<p><a href="https://github.com/DerrickWood/kraken2/blob/master/docs/MANUAL.markdown#system-requirements">Kraken System Requirements</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fnref:ksr" title="Jump back to footnote 2 in the text">↩</a></p>
</li>
<li id="fn:err">
<p><a href="ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR103/077/ERR10359977/ERR10359977.fastq.gz">ERR10359977.fastq.gz</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fnref:err" title="Jump back to footnote 3 in the text">↩</a></p>
</li>
<li id="fn:k2p">
<p><a href="https://benlangmead.github.io/aws-indexes/k2">Genomic Index Zone - k2</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fnref:k2p" title="Jump back to footnote 4 in the text">↩</a></p>
</li>
<li id="fn:vmt">
<p><a href="https://hoytech.com/vmtouch/">https://hoytech.com/vmtouch/</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fnref:vmt" title="Jump back to footnote 5 in the text">↩</a></p>
</li>
</ol>
</div>
    </div>
    </article><article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="../2024/07/mastering-kraken2-initial-runs.html" class="u-url">Mastering Kraken2 - Part 1 - Initial Runs</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                Chillar Anand
            </span></p>
            <p class="dateline">
            <a href="../2024/07/mastering-kraken2-initial-runs.html" rel="bookmark">
            <time class="published dt-published" datetime="2024-07-28T10:44:25+05:30" itemprop="datePublished" title="2024-07-28">2024-07-28</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <h4>Mastering Kraken2</h4>
<p><a href="../2024/07/mastering-kraken2-initial-runs.html">Part 1 - Initial Runs</a> (this post)</p>
<p><a href="../2024/07/mastering-kraken2-performance-optimisation.html">Part 2 - Performance Optimisation</a></p>
<p><a href="../2024/07/mastering-kraken2-build-custom-db.html">Part 3 - Building custom databases</a></p>
<p>Part 4 - Regular vs Fast Builds (upcoming)</p>
<p>Part 5 - Benchmarking (upcoming)</p>
<h4>Introduction</h4>
<p>Kraken2<sup id="fnref:Kraken2"><a class="footnote-ref" href="../2024/07/mastering-kraken2-initial-runs.html#fn:Kraken2">1</a></sup> is widely used for metagenomics taxonomic classification, and it has pre-built indexes for many organisms. In this series, we will learn</p>
<ul>
<li>How to set up kraken2, download pre-built indices</li>
<li>Run kraken2 (8GB RAM) at ~0.19 Mbp/m (million base pairs per minute)</li>
<li>Learn various ways to speed up the classification process</li>
<li>Run kraken2 (128GB RAM) at ~1200 Mbp/m</li>
<li>Build custom indices</li>
</ul>
<h4>Installation</h4>
<p>We can install kraken2 from source using the <code>install_kraken2.sh</code> script as per the manual<sup id="fnref:install_kraken2"><a class="footnote-ref" href="../2024/07/mastering-kraken2-initial-runs.html#fn:install_kraken2">2</a></sup>.</p>
<pre class="code literal-block">$ git clone https://github.com/DerrickWood/kraken2
$ <span class="nb">cd</span> kraken2
$ ./install_kraken2.sh /usr/local/bin
<span class="c1"># ensure kraken2 is in the PATH</span>
$ <span class="nb">export</span> <span class="nv">PATH</span><span class="o">=</span><span class="nv">$PATH</span>:/usr/local/bin
</pre>
<p>If you already have conda installed, you can install kraken2 from conda as well.</p>
<pre class="code literal-block">$ conda install -c bioconda kraken2
</pre>
<p>If you have <code>brew</code> installed on Linux or Mac(including M1), you can install kraken2 using <code>brew</code>.</p>
<pre class="code literal-block">$ brew install brewsci/bio/kraken2
</pre>
<h4>Download pre-built indices</h4>
<p>Building kraken2 indices take a lot of time and resources. For now, let's download and use the pre-built indices. In the final post, we will learn how to build the indices.</p>
<p>Genomic Index Zone<sup id="fnref:GenomicIndexZone"><a class="footnote-ref" href="../2024/07/mastering-kraken2-initial-runs.html#fn:GenomicIndexZone">3</a></sup> provides pre-built indices for kraken2. Let's download the standard database. It contains Refeq archaea, bacteria, viral, plasmid, human1, &amp; UniVec_Core. </p>
<pre class="code literal-block">$ wget https://genome-idx.s3.amazonaws.com/kraken/k2_standard_20240605.tar.gz
$ mkdir k2_standard
$ tar -xvf k2_standard_20240605.tar.gz -C k2_standard
</pre>
<p>The extracted directory contains three files - <code>hash.k2d</code>, <code>opts.k2d</code>, <code>taxo.k2d</code> which are the kraken2 database files.</p>
<pre class="code literal-block">$ ls -l *.k2d
.rw-r--r--  83G anand <span class="m">13</span> Jul <span class="m">12</span>:34 hash.k2d
.rw-r--r--   <span class="m">64</span> anand <span class="m">13</span> Jul <span class="m">12</span>:34 opts.k2d
.rw-r--r-- <span class="m">4</span>.0M anand <span class="m">13</span> Jul <span class="m">12</span>:34 taxo.k2d
</pre>
<h4>Classification</h4>
<p>To run the taxonomic classification, let's use <code>ERR10359977</code> human gut meta genome from NCBI SRA.</p>
<pre class="code literal-block">$ wget https://ftp.sra.ebi.ac.uk/vol1/fastq/ERR103/077/ERR10359977/ERR10359977.fastq.gz
$ kraken2 --db k2_standard --report report.txt ERR10359977.fastq.gz &gt; output.txt
</pre>
<p>By default, the machine I have used has 8GB RAM and an additioinal 8GB swap. Since kraken2 needs entire db(~80GB) in memory, when the process tries to consume more than 16GB memory, the kernel will kill the process. </p>
<pre class="code literal-block">$ <span class="nb">time</span> kraken2 --db k2_standard --paired SRR6915097_1.fastq.gz SRR6915097_2.fastq.gz &gt; output.txt
Loading database information...Command terminated by signal <span class="m">9</span>
<span class="m">0</span>.02user <span class="m">275</span>.83system <span class="m">8</span>:17.43elapsed <span class="m">55</span>%CPU 
</pre>
<p>To prevent this, let's increase the swap space to 128 GB.</p>
<pre class="code literal-block"><span class="c1"># Create an empty swapfile of 128GB</span>
sudo dd <span class="k">if</span><span class="o">=</span>/dev/zero <span class="nv">of</span><span class="o">=</span>/swapfile <span class="nv">bs</span><span class="o">=</span>1G <span class="nv">count</span><span class="o">=</span><span class="m">128</span>

<span class="c1"># Turn swap off - It might take several minutes</span>
sudo swapoff -a

<span class="c1"># Set the permissions for swapfile</span>
sudo chmod <span class="m">0600</span> /swapfile

<span class="c1"># make it a swap area</span>
sudo mkswap /swapfile  

<span class="c1"># Turn the swap on</span>
sudo swapon /swapfile
</pre>
<p>We can time the classification process using the <code>time</code> command.</p>
<pre class="code literal-block">$ <span class="nb">time</span> kraken2 --db k2_standard --report report.txt ERR10359977.fastq.gz &gt; output.txt
</pre>
<p>If you have a machine with large RAM, the same scenario can be simulated using <code>systemd-run</code>. This will limit the memory usage of kraken2 to 6.5GB. </p>
<pre class="code literal-block">$ <span class="nb">time</span> systemd-run --scope -p <span class="nv">MemoryMax</span><span class="o">=</span><span class="m">6</span>.5G --user <span class="nb">time</span> kraken2 --db k2_standard --report report.txt ERR10359977.fastq.gz &gt; output.txt
</pre>
<p>Depending on the CPU performance, this will take around ~40 minutes to complete.</p>
<pre class="code literal-block">Loading database information... <span class="k">done</span>.
<span class="m">95064</span> sequences <span class="o">(</span><span class="m">14</span>.35 Mbp<span class="o">)</span> processed <span class="k">in</span> <span class="m">1026</span>.994s <span class="o">(</span><span class="m">5</span>.6 Kseq/m, <span class="m">0</span>.84 Mbp/m<span class="o">)</span>.
  <span class="m">94939</span> sequences classified <span class="o">(</span><span class="m">99</span>.87%<span class="o">)</span>
  <span class="m">125</span> sequences unclassified <span class="o">(</span><span class="m">0</span>.13%<span class="o">)</span>
  <span class="m">4</span>.24user <span class="m">658</span>.68system <span class="m">38</span>:26.78elapsed <span class="m">28</span>%CPU 
</pre>
<p>If we try gut WGS(Whole Genome Sequence) sample like <code>SRR6915097</code> <sup id="fnref:srr1"><a class="footnote-ref" href="../2024/07/mastering-kraken2-initial-runs.html#fn:srr1">4</a></sup><sup id="fnref:srr2"><a class="footnote-ref" href="../2024/07/mastering-kraken2-initial-runs.html#fn:srr2">5</a></sup>. which contains ~3.3 Gbp, it will take weeks to complete.</p>
<pre class="code literal-block">$ wget -c https://ftp.sra.ebi.ac.uk/vol1/fastq/SRR691/007/SRR6915097/SRR6915097_1.fastq.gz
$ wget -c https://ftp.sra.ebi.ac.uk/vol1/fastq/SRR691/007/SRR6915097/SRR6915097_2.fastq.gz

$ <span class="nb">time</span> systemd-run --scope -p <span class="nv">MemoryMax</span><span class="o">=</span>6G --user <span class="nb">time</span> kraken2 --db k2_standard --paired SRR6915097_1.fastq.gz SRR6915097_2.fastq.gz &gt; output.txt
</pre>
<p>I tried running this on 8 GB machine. Even after 10 days, it processed only 10% of the data.</p>
<p>If we have to process a large number of such samples, it takes months and this is not a practical solution. </p>
<h4>Conclusion</h4>
<p>In this post, we ran kraken2 on an 8GB machine and learned that it is not feasible to run kraken2 on large samples.</p>
<p>In the next post, we will learn how to speed up the classification process and run classification at 1200 Mbp/m.</p>
<p><strong>Next</strong>: <a href="../2024/07/mastering-kraken2-performance-optimisation.html">Part 2 - Performance Optimisation</a></p>
<div class="footnote">
<hr>
<ol>
<li id="fn:Kraken2">
<p><a href="https://ccb.jhu.edu/software/kraken2/">Kraken2</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-initial-runs.html#fnref:Kraken2" title="Jump back to footnote 1 in the text">↩</a></p>
</li>
<li id="fn:install_kraken2">
<p><a href="https://github.com/DerrickWood/kraken2/blob/master/docs/MANUAL.markdown#installation">Kraken2 - Manual - Install</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-initial-runs.html#fnref:install_kraken2" title="Jump back to footnote 2 in the text">↩</a></p>
</li>
<li id="fn:GenomicIndexZone">
<p><a href="https://benlangmead.github.io/aws-indexes/k2">Genomic Index Zone - k2</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-initial-runs.html#fnref:GenomicIndexZone" title="Jump back to footnote 3 in the text">↩</a></p>
</li>
<li id="fn:srr1">
<p><a href="https://ftp.sra.ebi.ac.uk/vol1/fastq/SRR691/007/SRR6915097/SRR6915097_1.fastq.gz">SRR6915097_1.fastq.gz</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-initial-runs.html#fnref:srr1" title="Jump back to footnote 4 in the text">↩</a></p>
</li>
<li id="fn:srr2">
<p><a href="https://ftp.sra.ebi.ac.uk/vol1/fastq/SRR691/007/SRR6915097/SRR6915097_2.fastq.gz">SRR6915097_1.fastq.gz</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-initial-runs.html#fnref:srr2" title="Jump back to footnote 5 in the text">↩</a></p>
</li>
</ol>
</div>
    </div>
    </article><article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="../2024/06/headlamp-k8s-lens-open-source-alternative.html" class="u-url">Headlamp - k8s Lens open source alternative</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                Chillar Anand
            </span></p>
            <p class="dateline">
            <a href="../2024/06/headlamp-k8s-lens-open-source-alternative.html" rel="bookmark">
            <time class="published dt-published" datetime="2024-06-24T01:48:02+05:30" itemprop="datePublished" title="2024-06-24">2024-06-24</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <p><img alt="headlamp - Open source Kubernetes Lens alternator" src="../images/headlamp-k8s-lens-open-source-alternative.png"></p>
<p>Since Lens is not open source, I tried out monokle, octant, k9s, and headlamp<sup id="fnref:headlamp"><a class="footnote-ref" href="../2024/06/headlamp-k8s-lens-open-source-alternative.html#fn:headlamp">1</a></sup>. Among them, headlamp UI &amp; features are closest to Lens. </p>
<h4>Headlamp</h4>
<p>Headlamp is CNCF sandbox project that provides cross-platform desktop application to manage Kubernetes clusters. It auto-detects clusters and provides cluster wide resource usage by default. </p>
<p>It can also be installed inside the cluster and can be accessed using a web browser. This is useful when we want to access the cluster from a mobile device.</p>
<pre class="code literal-block">$ helm repo add headlamp https://headlamp-k8s.github.io/headlamp/

$ helm install headlamp headlamp/headlamp
</pre>
<p>Lets port-forward the service &amp; copy the token to access it.</p>
<pre class="code literal-block">$ kubectl create token headlamp

<span class="c1"># we can do this via headlamp UI as well</span>
$ kubectl port-forward service/headlamp <span class="m">8080</span>:80
</pre>
<p>Now, we can access the headlamp UI at <a href="http://">http://localhost:8080</a>.</p>
<p><img alt="headlamp - Open source Kubernetes Lens alternator" src="../images/headlamp-k8s-lens-open-source-alternative2.png"></p>
<h4>Conclusion</h4>
<p>If you are looking for an open source alternative to Lens, headlamp is a good choice. It provides a similar UI &amp; features as Lens, and it is accessible via mobile devices as well. </p>
<div class="footnote">
<hr>
<ol>
<li id="fn:headlamp">
<p><a href="https://headlamp.dev/">https://headlamp.dev/</a> <a class="footnote-backref" href="../2024/06/headlamp-k8s-lens-open-source-alternative.html#fnref:headlamp" title="Jump back to footnote 1 in the text">↩</a></p>
</li>
</ol>
</div>
    </div>
    </article><article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="../2024/06/macos-log-track-cpu-ram-usage.html" class="u-url">macOS - Log &amp; track historical CPU, RAM usage</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                Chillar Anand
            </span></p>
            <p class="dateline">
            <a href="../2024/06/macos-log-track-cpu-ram-usage.html" rel="bookmark">
            <time class="published dt-published" datetime="2024-06-01T01:48:02+05:30" itemprop="datePublished" title="2024-06-01">2024-06-01</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <p><img alt="macOS - Log CPU &amp; RAM history" src="../images/mac-log-cpu-ram-grafana.png"></p>
<p>In macOS, we can use inbuilt <code>Activity Monitor</code> or third party apps like <code>Stats</code> to check the live CPU/RAM usage. But, we can't track the historical CPU &amp; memory usage. <code>sar</code>, <code>atop</code> can track the historical CPU &amp; memory usage. But, they are not available for macOS.</p>
<h4>Netdata</h4>
<p>Netdata<sup id="fnref:Netdata"><a class="footnote-ref" href="../2024/06/macos-log-track-cpu-ram-usage.html#fn:Netdata">1</a></sup> is an open source observability tool that can monitor CPU, RAM, network, disk usage. It can also track the historical data. </p>
<p>Unfortunately, it is not stable on macOS. I tried installing it on multiple macbooks, but it didn't work. I raised an issue<sup id="fnref:netdata_issue"><a class="footnote-ref" href="../2024/06/macos-log-track-cpu-ram-usage.html#fn:netdata_issue">2</a></sup> on their GitHub repository and the team mentioned that macOS is a low priority for them.</p>
<h4>Glances</h4>
<p>Glances<sup id="fnref:Glances"><a class="footnote-ref" href="../2024/06/macos-log-track-cpu-ram-usage.html#fn:Glances">3</a></sup> is a cross-platform monitoring tool that can monitor CPU, RAM, network, disk usage. It can also track the historical data.</p>
<p>We can install it using Brew or pip.</p>
<pre class="code literal-block">$ brew install glances

$ pip install glances
</pre>
<p>Once it is installed, we can monitor the resource usage using the below command.</p>
<pre class="code literal-block">$ glances
</pre>
<p><img alt="macOS - Log CPU &amp; RAM history" src="../images/mac-log-cpu-ram-glances.png"></p>
<p>Glances can log historical data to a file using the below command.</p>
<pre class="code literal-block">$ glances --export-csv /tmp/glances.csv
</pre>
<p>In addition to that, it can log data to services like influxdb, prometheus, etc.</p>
<p>Let's install influxdb and export stats to it.</p>
<pre class="code literal-block">$ brew install influxdb
$ brew services start influxdb
$ influx setup

$ python -m pip install influxdb-client

$ cat glances.conf
<span class="o">[</span>influxdb<span class="o">]</span>
<span class="nv">host</span><span class="o">=</span>localhost
<span class="nv">port</span><span class="o">=</span><span class="m">8086</span>
<span class="nv">protocol</span><span class="o">=</span>http
<span class="nv">org</span><span class="o">=</span>avilpage
<span class="nv">bucket</span><span class="o">=</span>glances
<span class="nv">token</span><span class="o">=</span>secret_token

$ glances --export-influxdb -C glances.conf
</pre>
<p>We can view stats in the influxdb from Data Explorer web UI at <a href="http://localhost:8086">http://localhost:8086</a>.</p>
<p><img alt="macOS - Log CPU &amp; RAM history" src="../images/mac-log-cpu-ram-influxdb.png"></p>
<p>Glances provides a prebuilt Grafana dashboard<sup id="fnref:grafana_dashboard"><a class="footnote-ref" href="../2024/06/macos-log-track-cpu-ram-usage.html#fn:grafana_dashboard">4</a></sup> that we can import to visualize the stats. </p>
<p>From Grafana -&gt; Dashboard -&gt; Import, we can import the dashboard using the above URL.</p>
<p><img alt="macOS - Log CPU &amp; RAM history" src="../images/mac-log-cpu-ram-grafana.png"></p>
<h4>Conclusion</h4>
<p>In addition to InfluxDB, Glances can export data to ~20 services. So far, it is the best tool to log, track and view historical CPU, RAM, network and disk usage in macOS. The same method works for Linux and Windows as well.</p>
<div class="footnote">
<hr>
<ol>
<li id="fn:Netdata">
<p><a href="https://github.com/netdata/netdata">https://github.com/netdata/netdata</a> <a class="footnote-backref" href="../2024/06/macos-log-track-cpu-ram-usage.html#fnref:Netdata" title="Jump back to footnote 1 in the text">↩</a></p>
</li>
<li id="fn:netdata_issue">
<p><a href="https://github.com/netdata/netdata/issues/16696">https://github.com/netdata/netdata/issues/16696</a> <a class="footnote-backref" href="../2024/06/macos-log-track-cpu-ram-usage.html#fnref:netdata_issue" title="Jump back to footnote 2 in the text">↩</a></p>
</li>
<li id="fn:Glances">
<p><a href="https://github.com/nicolargo/glances">https://github.com/niolargo/glances</a> <a class="footnote-backref" href="../2024/06/macos-log-track-cpu-ram-usage.html#fnref:Glances" title="Jump back to footnote 3 in the text">↩</a></p>
</li>
<li id="fn:grafana_dashboard">
<p><a href="https://glances.readthedocs.io/en/latest/gw/influxdb.html#grafana">https://glances.readthedocs.io/en/latest/gw/influxdb.html#grafana</a> <a class="footnote-backref" href="../2024/06/macos-log-track-cpu-ram-usage.html#fnref:grafana_dashboard" title="Jump back to footnote 4 in the text">↩</a></p>
</li>
</ol>
</div>
    </div>
    </article><article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="../2024/05/managing-zscaler-connectivity-from-command-line.html" class="u-url">Automating Zscaler Connectivity on macOS</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                Chillar Anand
            </span></p>
            <p class="dateline">
            <a href="../2024/05/managing-zscaler-connectivity-from-command-line.html" rel="bookmark">
            <time class="published dt-published" datetime="2024-05-14T07:49:53+05:30" itemprop="datePublished" title="2024-05-14">2024-05-14</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <h4>Introduction</h4>
<p>Zscaler is a cloud-based security service that provides secure internet access via VPN. Unfortunately, Zscaler does not provide a command-line interface to connect to the VPN. We can't use AppleScript to automate the connectivity as well.</p>
<h4>Automating Zscaler Connectivity</h4>
<p>Once Zscaler is installed on macOS, if we search for <code>LaunchAgents</code> &amp; <code>LaunchDaemons</code> directories, we can find the Zscaler plist files. </p>
<pre class="code literal-block">$ sudo find /Library/LaunchAgents -name <span class="s1">'*zscaler*'</span>
/Library/LaunchAgents/com.zscaler.tray.plist


$ sudo find /Library/LaunchDaemons -name <span class="s1">'*zscaler*'</span>
/Library/LaunchDaemons/com.zscaler.service.plist
/Library/LaunchDaemons/com.zscaler.tunnel.plist
/Library/LaunchDaemons/com.zscaler.UPMServiceController.plist
</pre>
<p>To connect to Zscaler, we can load these services.</p>
<pre class="code literal-block"><span class="ch">#!/bin/bash</span>

/usr/bin/open -a /Applications/Zscaler/Zscaler.app --hide
sudo find /Library/LaunchAgents -name <span class="s1">'*zscaler*'</span> -exec launchctl load <span class="o">{}</span> <span class="se">\;</span>
sudo find /Library/LaunchDaemons -name <span class="s1">'*zscaler*'</span> -exec launchctl load <span class="o">{}</span> <span class="se">\;</span>
</pre>
<p>To disconnect from Zscaler, we can unload all of them.</p>
<pre class="code literal-block"><span class="ch">#!/bin/bash</span>

sudo find /Library/LaunchAgents -name <span class="s1">'*zscaler*'</span> -exec launchctl unload <span class="o">{}</span> <span class="se">\;</span>
sudo find /Library/LaunchDaemons -name <span class="s1">'*zscaler*'</span> -exec launchctl unload <span class="o">{}</span> <span class="se">\;</span>
</pre>
<p>To automatically toggle the connectivity, we can create a shell script.</p>
<pre class="code literal-block"><span class="ch">#!/bin/bash</span>

<span class="k">if</span> <span class="o">[[</span> <span class="k">$(</span>pgrep -x Zscaler<span class="k">)</span> <span class="o">]]</span><span class="p">;</span> <span class="k">then</span>
    <span class="nb">echo</span> <span class="s2">"Disconnecting from Zscaler"</span>
    sudo find /Library/LaunchAgents -name <span class="s1">'*zscaler*'</span> -exec launchctl unload <span class="o">{}</span> <span class="se">\;</span>
    sudo find /Library/LaunchDaemons -name <span class="s1">'*zscaler*'</span> -exec launchctl unload <span class="o">{}</span> <span class="se">\;</span>
<span class="k">else</span>
    <span class="nb">echo</span> <span class="s2">"Connecting to Zscaler"</span>
    /usr/bin/open -a /Applications/Zscaler/Zscaler.app --hide
    sudo find /Library/LaunchAgents -name <span class="s1">'*zscaler*'</span> -exec launchctl load <span class="o">{}</span> <span class="se">\;</span>
    sudo find /Library/LaunchDaemons -name <span class="s1">'*zscaler*'</span> -exec launchctl load <span class="o">{}</span> <span class="se">\;</span>
<span class="k">fi</span>
</pre>
<p>Raycast is an alternative to default spotlight search on macOS. We can create a script to toggle connectivity to Zscaler.</p>
<pre class="code literal-block"><span class="ch">#!/bin/bash</span>

<span class="c1"># Required parameters:</span>
<span class="c1"># @raycast.schemaVersion 1</span>
<span class="c1"># @raycast.title toggle zscaler</span>
<span class="c1"># @raycast.mode silent</span>

<span class="c1"># Optional parameters:</span>
<span class="c1"># @raycast.icon ☁️</span>

<span class="c1"># Documentation:</span>
<span class="c1"># @raycast.author chillaranand</span>
<span class="c1"># @raycast.authorURL https://avilpage.com/</span>

<span class="k">if</span> <span class="o">[[</span> <span class="k">$(</span>pgrep -x Zscaler<span class="k">)</span> <span class="o">]]</span><span class="p">;</span> <span class="k">then</span>
    <span class="nb">echo</span> <span class="s2">"Disconnecting from Zscaler"</span>
    sudo find /Library/LaunchAgents -name <span class="s1">'*zscaler*'</span> -exec launchctl unload <span class="o">{}</span> <span class="se">\;</span>
    sudo find /Library/LaunchDaemons -name <span class="s1">'*zscaler*'</span> -exec launchctl unload <span class="o">{}</span> <span class="se">\;</span>
<span class="k">else</span>
    <span class="nb">echo</span> <span class="s2">"Connecting to Zscaler"</span>
    /usr/bin/open -a /Applications/Zscaler/Zscaler.app --hide
    sudo find /Library/LaunchAgents -name <span class="s1">'*zscaler*'</span> -exec launchctl load <span class="o">{}</span> <span class="se">\;</span>
    sudo find /Library/LaunchDaemons -name <span class="s1">'*zscaler*'</span> -exec launchctl load <span class="o">{}</span> <span class="se">\;</span>
<span class="k">fi</span>
</pre>
<p>Save this script to a folder. From <code>Raycast Settings</code> -&gt; <code>Extensions</code> -&gt; <code>Add Script Directory</code>, we can select this folder, and the script will be available in Raycast. </p>
<p><img alt="raycast-connect-toggle" src="../images/automate-zscaler-connect2.png"></p>
<p>We can assign a shortcut key to the script for quick access.</p>
<p><img alt="raycast-connect-toggle" src="../images/automate-zscaler-connect.png"></p>
<h4>Conclusion</h4>
<p>Even though Zscaler does not provide a command-line interface, we can automate the connectivity using the above scripts.</p>
    </div>
    </article><article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="../2024/05/screen-time-alerts-from-activity-watch.html" class="u-url">Screen Time Alerts from Activity Watch</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                Chillar Anand
            </span></p>
            <p class="dateline">
            <a href="../2024/05/screen-time-alerts-from-activity-watch.html" rel="bookmark">
            <time class="published dt-published" datetime="2024-05-01T06:59:27+05:30" itemprop="datePublished" title="2024-05-01">2024-05-01</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <h4>Introduction</h4>
<div class="embed-responsive embed-responsive-16by9">
<iframe class="embed-responsive-item" src="https://www.youtube.com/embed/41A0r2H-EXs" allowfullscreen>
</iframe>
</div>

<p><br></p>
<p>Activity Watch<sup id="fnref:activity_watch"><a class="footnote-ref" href="../2024/05/screen-time-alerts-from-activity-watch.html#fn:activity_watch">1</a></sup> is a cross-platform open-source time-tracking tool that helps us to track time spent on applications and websites.</p>
<p><img alt="Activity Watch" src="../images/activity-watch-alerts.png"></p>
<p>At the moment, Activity Watch doesn't have any feature to show screen time alerts. In this post, we will see how to show screen time alerts using Activity Watch.</p>
<h4>Python Script</h4>
<p>Activity Watch provides an API to interact with the Activity Watch server. We can use the API to get the screen time data and show alerts.</p>
<pre class="code literal-block"><span class="kn">import</span> <span class="nn">json</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">from</span> <span class="nn">datetime</span> <span class="kn">import</span> <span class="n">datetime</span>

<span class="kn">import</span> <span class="nn">requests</span>


<span class="k">def</span> <span class="nf">get_nonafk_events</span><span class="p">(</span><span class="n">timeperiods</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="n">headers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">"Content-type"</span><span class="p">:</span> <span class="s2">"application/json"</span><span class="p">,</span> <span class="s2">"charset"</span><span class="p">:</span> <span class="s2">"utf-8"</span><span class="p">}</span>
    <span class="n">query</span> <span class="o">=</span> <span class="s2">"""afk_events = query_bucket(find_bucket('aw-watcher-afk_'));</span>
<span class="s2">window_events = query_bucket(find_bucket('aw-watcher-window_'));</span>
<span class="s2">window_events = filter_period_intersect(window_events, filter_keyvals(afk_events, 'status', ['not-afk']));</span>
<span class="s2">RETURN = merge_events_by_keys(window_events, ['app', 'title']);"""</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span><span class="p">)</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">"timeperiods"</span><span class="p">:</span> <span class="n">timeperiods</span><span class="p">,</span>
        <span class="s2">"query"</span><span class="p">:</span> <span class="n">query</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">r</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">post</span><span class="p">(</span>
        <span class="s2">"http://localhost:5600/api/0/query/"</span><span class="p">,</span>
        <span class="n">data</span><span class="o">=</span><span class="nb">bytes</span><span class="p">(</span><span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">data</span><span class="p">),</span> <span class="s2">"utf-8"</span><span class="p">),</span>
        <span class="n">headers</span><span class="o">=</span><span class="n">headers</span><span class="p">,</span>
        <span class="n">params</span><span class="o">=</span><span class="p">{},</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">text</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>


<span class="k">def</span> <span class="nf">main</span><span class="p">():</span>
    <span class="n">now</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span>
    <span class="n">timeperiods</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">"/"</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">now</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">hour</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">minute</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">second</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">isoformat</span><span class="p">(),</span> <span class="n">now</span><span class="o">.</span><span class="n">isoformat</span><span class="p">()])</span>
    <span class="p">]</span>
    <span class="n">events</span> <span class="o">=</span> <span class="n">get_nonafk_events</span><span class="p">(</span><span class="n">timeperiods</span><span class="p">)</span>

    <span class="n">total_time_secs</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">event</span><span class="p">[</span><span class="s2">"duration"</span><span class="p">]</span> <span class="k">for</span> <span class="n">event</span> <span class="ow">in</span> <span class="n">events</span><span class="p">])</span>
    <span class="n">total_time_mins</span> <span class="o">=</span> <span class="n">total_time_secs</span> <span class="o">/</span> <span class="mi">60</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Total time: </span><span class="si">{</span><span class="n">total_time_mins</span><span class="si">}</span><span class="s2"> seconds"</span><span class="p">)</span>
    <span class="n">hours</span><span class="p">,</span> <span class="n">minutes</span> <span class="o">=</span> <span class="nb">divmod</span><span class="p">(</span><span class="n">total_time_mins</span><span class="p">,</span> <span class="mi">60</span><span class="p">)</span>
    <span class="n">minutes</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">minutes</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Screen Time: </span><span class="si">{</span><span class="n">hours</span><span class="si">}</span><span class="s2"> hours </span><span class="si">{</span><span class="n">minutes</span><span class="si">}</span><span class="s2"> minutes"</span><span class="p">)</span>

    <span class="c1"># show mac notification</span>
    <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="sa">f</span><span class="s2">"osascript -e 'display notification </span><span class="se">\"</span><span class="si">{</span><span class="n">hours</span><span class="si">}</span><span class="s2"> hours </span><span class="si">{</span><span class="n">minutes</span><span class="si">}</span><span class="s2"> minutes</span><span class="se">\"</span><span class="s2"> with title </span><span class="se">\"</span><span class="s2">Screen TIme</span><span class="se">\"</span><span class="s2">'"</span><span class="p">)</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">"__main__"</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre>
<p>This script<sup id="fnref:github"><a class="footnote-ref" href="../2024/05/screen-time-alerts-from-activity-watch.html#fn:github">2</a></sup> will show the screen time alerts using the Activity Watch API. We can run this script using the below command.</p>
<pre class="code literal-block">$ python screen_time_alerts.py
</pre>
<p><img alt="Screen Time Alerts" src="../images/activity-watch-alerts2.png"></p>
<p>We can set up a cron job to run this script every hour to show screen time alerts.</p>
<pre class="code literal-block">$ crontab -e
<span class="m">0</span> * * * * python screen_time_alerts.py
</pre>
<p>We can also modify the script to show alerts only when the screen time exceeds a certain limit.</p>
<h4>Conclusion</h4>
<p>Since Actvity Watch is open-source and provides an API, we can extend its functionality to show screen time alerts. We can also use the API to create custom reports and dashboards.</p>
<div class="footnote">
<hr>
<ol>
<li id="fn:activity_watch">
<p><a href="https://activitywatch.net/">Activity Watch</a> <a class="footnote-backref" href="../2024/05/screen-time-alerts-from-activity-watch.html#fnref:activity_watch" title="Jump back to footnote 1 in the text">↩</a></p>
</li>
<li id="fn:github">
<p><a href="https://github.com/ChillarAnand/avilpage.com/blob/master/scripts/activity_watch_alerts.py">Screen Time Alerts Script</a> <a class="footnote-backref" href="../2024/05/screen-time-alerts-from-activity-watch.html#fnref:github" title="Jump back to footnote 2 in the text">↩</a></p>
</li>
</ol>
</div>
    </div>
    </article><article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="../2024/04/how-to-setup-ftp-server-on-mac-os.html" class="u-url">Setup FTP server on Mac OS X</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                Chillar Anand
            </span></p>
            <p class="dateline">
            <a href="../2024/04/how-to-setup-ftp-server-on-mac-os.html" rel="bookmark">
            <time class="published dt-published" datetime="2024-04-30T23:34:14+05:30" itemprop="datePublished" title="2024-04-30">2024-04-30</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <div class="embed-responsive embed-responsive-16by9">
<iframe class="embed-responsive-item" src="https://www.youtube.com/embed/WmmB-6MXRYc" allowfullscreen>
</iframe>
</div>

<p><br></p>
<p>On Linux &amp; Mac OS X, Python comes pre-installed. On Windows, we can install it from Windows store or from <a href="https://python.org">https://python.org</a> website.</p>
<p>We can verify the Python version using the below command.</p>
<pre class="code literal-block">$ python --version
Python <span class="m">3</span>.11.6
</pre>
<p>We can use the <code>pyftpdlib</code> library to create an FTP server. We can install the library using the below command.</p>
<pre class="code literal-block">$ python -m pip install pyftpdlib
<span class="o">[</span>I <span class="m">11</span>:28:21<span class="o">]</span> concurrency model: async
<span class="o">[</span>I <span class="m">11</span>:28:21<span class="o">]</span> masquerade <span class="o">(</span>NAT<span class="o">)</span> address: None
<span class="o">[</span>I <span class="m">11</span>:28:21<span class="o">]</span> passive ports: None
<span class="o">[</span>I <span class="m">11</span>:28:21<span class="o">]</span> &gt;&gt;&gt; starting FTP server on :::2121, <span class="nv">pid</span><span class="o">=</span><span class="m">99951</span> <span class="o">&lt;&lt;&lt;</span>
</pre>
<p>Now, we can start the FTP server using the below command.</p>
<pre class="code literal-block">$ python -m pyftpdlib
</pre>
<p>It will start the FTP server on port 2121. We can connect to the FTP server using the below command.</p>
<pre class="code literal-block">$ ftp localhost <span class="m">2121</span>
</pre>
    </div>
    </article><article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="../2024/04/timestamp-to-relative-time-kibana-scripted-fields.html" class="u-url">Timestamp to Relative Time - Kibana Scripted fields</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                Chillar Anand
            </span></p>
            <p class="dateline">
            <a href="../2024/04/timestamp-to-relative-time-kibana-scripted-fields.html" rel="bookmark">
            <time class="published dt-published" datetime="2024-04-10T14:15:33+05:30" itemprop="datePublished" title="2024-04-10">2024-04-10</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <p>When browsing logs in Kibana, there will be a timestamp stamp field on the left for all the docs. It is difficult to read &amp; comprehend the timestamp in the logs. It would be better if we can convert the timestamp to a human-readable relative time like <code>5 minutes ago</code>, <code>1 hour ago</code>, etc.</p>
<h4>Kibana Scripted Fields</h4>
<p>Kibana provides a feature called scripted fields to create new fields in the index pattern. We can use this feature to convert the timestamp to a relative time.</p>
<p><img alt="kibana-relative-time" src="../images/kibana-relative-time1.png"></p>
<p>Go to <code>Stack Management</code> -&gt; <code>Index Patterns</code> -&gt; <code>Create index pattern</code> -&gt; Select the index pattern -&gt; <code>Scripted fields</code>, click on <code>Add scripted field</code>, add the below script.</p>
<pre class="code literal-block"><span class="n">long</span><span class="w"> </span><span class="n">now</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="k">new</span><span class="w"> </span><span class="nc">Date</span><span class="p">().</span><span class="n">getTime</span><span class="p">();</span><span class="w"></span>

<span class="n">long</span><span class="w"> </span><span class="nc">timestamp</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">doc</span><span class="o">[</span><span class="n">'@timestamp'</span><span class="o">]</span><span class="p">.</span><span class="k">value</span><span class="p">.</span><span class="n">toInstant</span><span class="p">().</span><span class="n">toEpochMilli</span><span class="p">();</span><span class="w"></span>
<span class="n">long</span><span class="w"> </span><span class="n">diff</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">now</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="nc">timestamp</span><span class="p">;</span><span class="w"></span>
<span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="n">diff</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="mi">7200000</span><span class="p">)</span><span class="w"> </span><span class="err">{</span><span class="w"></span>
<span class="w">  </span><span class="k">return</span><span class="w"> </span><span class="n">Math</span><span class="p">.</span><span class="nf">round</span><span class="p">(</span><span class="n">diff</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">3600000</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="ss">" hours ago"</span><span class="p">;</span><span class="w"></span>
<span class="err">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="n">diff</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="mi">3600000</span><span class="p">)</span><span class="w"> </span><span class="err">{</span><span class="w"></span>
<span class="w">  </span><span class="k">return</span><span class="w"> </span><span class="n">Math</span><span class="p">.</span><span class="nf">round</span><span class="p">(</span><span class="n">diff</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">3600000</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="ss">" hour ago"</span><span class="p">;</span><span class="w"></span>
<span class="err">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="n">diff</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="mi">120000</span><span class="p">)</span><span class="w"> </span><span class="err">{</span><span class="w"></span>
<span class="w">  </span><span class="k">return</span><span class="w"> </span><span class="n">Math</span><span class="p">.</span><span class="nf">round</span><span class="p">(</span><span class="n">diff</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">60000</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="ss">" minutes ago"</span><span class="p">;</span><span class="w"></span>
<span class="err">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="n">diff</span><span class="w"> </span><span class="o">&gt;</span><span class="w"> </span><span class="mi">60000</span><span class="p">)</span><span class="w"> </span><span class="err">{</span><span class="w"></span>
<span class="w">  </span><span class="k">return</span><span class="w"> </span><span class="p">(</span><span class="n">Math</span><span class="p">.</span><span class="nf">round</span><span class="p">(</span><span class="n">diff</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">60000</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="ss">" minute ago"</span><span class="p">);</span><span class="w"></span>
<span class="err">}</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="err">{</span><span class="w"></span>
<span class="w">  </span><span class="k">return</span><span class="w"> </span><span class="n">Math</span><span class="p">.</span><span class="nf">round</span><span class="p">(</span><span class="n">diff</span><span class="w"> </span><span class="o">/</span><span class="w"> </span><span class="mi">1000</span><span class="p">)</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="ss">" seconds ago"</span><span class="p">;</span><span class="w"></span>
<span class="err">}</span><span class="w"></span>
</pre>
<p>Once the field is saved, we can go back to <code>Discover</code> and see the new field in the logs. We can toggle the visibility of the <code>Relative Time</code> field to see the relative time.</p>
<p><img alt="kibana-relative-time" src="../images/kibana-relative-time2.png"></p>
<h4>Conclusion</h4>
<p>Instead of looking at the timestamp and calculating the relative time in our head, we can use relative time in Kibana . This will make it easier to read &amp; comprehend the logs.</p>
    </div>
    </article><article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="../2024/03/the-strange-case-of-mac-and-linux.html" class="u-url">The Strange Case of Dr. Linux and Mr. Mac</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                Chillar Anand
            </span></p>
            <p class="dateline">
            <a href="../2024/03/the-strange-case-of-mac-and-linux.html" rel="bookmark">
            <time class="published dt-published" datetime="2024-03-30T09:53:53+05:30" itemprop="datePublished" title="2024-03-30">2024-03-30</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <p>Few days back, some of the tests started failing on CI server. When I tried to run the tests locally, they were passing.</p>
<p>After debugging for a while, I found that the tests were failing because of the case sensitivity of the file system. One of the developer was using Linux and had committed 2 files with the same name but different case(<code>config.json</code>, <code>Config.json</code>).</p>
<p>Linux file system is case-sensitive. So these 2 files will be shown as 2 different files.</p>
<p><img alt="linux-file-system" src="../images/linux-git-case-sensitive.png"></p>
<p>But Mac/Windows file system is case-insensitive. Out of these 2 files, only one file will be shown.</p>
<p><img alt="mac-file-system" src="../images/mac-git-case-insensitive.png"></p>
<p>Due to this, the tests were failing on Linux but passing on Mac. Once the case of the file was corrected, the tests started passing on both the systems.</p>
<p>I have been using Mac for a long time and never faced this issue. Even though Mac's APFS is case-insensitive, we can create a case-sensitive volume using Disk Utility. </p>
<p><img alt="case-sensitive-volume" src="../images/mac-case-sensitive-volume.png"></p>
<p>We have to be aware of these differences when working on a project with developers using different OS.</p>
    </div>
    </article>
</div>
        <ul class="pager postindexpager clearfix">
<li class="next"><a href="index-19.html" rel="next">Older posts</a></li>
        </ul>
<!--End of body content--><footer id="footer"><div class="container align-items-center justify-content-center d-flex">

<footer class="footer"><a href="https://github.com/ChillarAnand">
<img src="../images/icons8-github.svg" alt="github-chillar-anand" height="34"></a>

  

<a href="https://stackoverflow.com/users/2698552/chillar-anand">
<img src="../images/icons8-so.svg" alt="github-chillar-anand" height="30"></a>

  

<a href="https://youtube.com/@avilpage">
<img src="../images/icons8-youtube.svg" alt="github-chillar-anand" height="34"></a>

  

<a href="https://linkedin.com/in/chillaranand">
<img src="../images/icons8-linkedin.svg" alt="github-chillar-anand" height="34"></a>

  

<a href="https://twitter.com/chillaranand">
<img src="../images/icons8-twitter.svg" alt="github-chillar-anand" height="34"></a>

</footer>
</div>

<br><br></footer>
</div>
    </div>
    </div>

                <script src="../assets/js/jquery.min.js"></script><script src="../assets/js/popper.min.js"></script><script src="../assets/js/bootstrap.min.js"></script><script src="../assets/js/baguetteBox.min.js"></script><script>
    baguetteBox.run('div#content', {
        ignoreClass: 'islink',
        captions: function(element){var i=element.getElementsByTagName('img')[0];return i===undefined?'':i.alt;}});
    </script>
</body>
</html>
