<!DOCTYPE html>
<html prefix="
og: http://ogp.me/ns# article: http://ogp.me/ns/article#
" lang="en">
<head>
<meta charset="utf-8">
<meta name="description" content="Avil Page - Personal &amp; tech blog by Pandikunta Anand Reddy (aka Chillar Anand)">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Avil Page (old posts, page 20) | Avil Page</title>
<link href="../assets/css/bootstrap.min.css" rel="stylesheet" type="text/css">
<link href="../assets/css/baguetteBox.min.css" rel="stylesheet" type="text/css">
<link href="../assets/css/rst.css" rel="stylesheet" type="text/css">
<link href="../assets/css/code.css" rel="stylesheet" type="text/css">
<link href="../assets/css/theme.css" rel="stylesheet" type="text/css">
<meta name="theme-color" content="#5670d4">
<meta name="generator" content="Nikola (getnikola.com)">
<link rel="alternate" type="application/rss+xml" title="RSS" hreflang="en" href="../rss.xml">
<link rel="canonical" href="https://avilpage.com/blog/index-20.html">
<link rel="prev" href="." type="text/html">
<link rel="next" href="index-19.html" type="text/html">
<!--[if lt IE 9]><script src="../assets/js/html5.js"></script><![endif]--><style>

 h1 {
     font-weight: bold;
     font-size: 25px;
 }

</style>
</head>
<body>
    <a href="#content" class="sr-only sr-only-focusable">Skip to main content</a>

    <!-- Menubar -->
    <nav class="navbar navbar-expand-md static-top mb-4
                navbar-dark
                bg-dark
                "><div class="container">
<!-- This keeps the margins nice -->
            <a class="navbar-brand" href="../">

                    <span id="blog-title">Avil Page</span>
            </a>
            <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#bs-navbar" aria-controls="bs-navbar" aria-expanded="false" aria-label="Toggle navigation">
                <span class="navbar-toggler-icon"></span>
            </button>

            <div class="collapse navbar-collapse" id="bs-navbar">
                <ul class="navbar-nav mr-auto">
<li class="nav-item">
<a href="../guides.html" class="nav-link">Guides</a>
                </li>
<li class="nav-item">
<a href="../archive.html" class="nav-link">Blog</a>
                </li>
<li class="nav-item">
<a href="../books.html" class="nav-link">Books</a>
                </li>
<li class="nav-item">
<a href="../talks.html" class="nav-link">Talks</a>
                </li>
<li class="nav-item">
<a href="../projects.html" class="nav-link">Projects</a>
                </li>
<li class="nav-item">
<a href="../rss.xml" class="nav-link">RSS</a>

                    
                </li>
</ul>
<ul class="navbar-nav navbar-right"></ul>
</div>
<!-- /.navbar-collapse -->
        </div>
<!-- /.container -->
    </nav><!-- End of Menubar --><div class="container" id="content" role="main">
    <div class="row justify-content-md-center">
        <div class="body-content col col-lg-6">
            <!--Body content-->
            
    
    
<div class="postindex">
    <article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="../2024/07/mastering-kraken2-build-custom-db.html" class="u-url">Mastering Kraken2 - Part 3 - Build Custom Database</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                Chillar Anand
            </span></p>
            <p class="dateline">
            <a href="../2024/07/mastering-kraken2-build-custom-db.html" rel="bookmark">
            <time class="published dt-published" datetime="2024-08-01T10:52:30+05:30" itemprop="datePublished" title="2024-08-01">2024-08-01</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <h4>Mastering Kraken2</h4>
<p><a href="../2024/07/mastering-kraken2-initial-runs.html">Part 1 - Initial Runs</a></p>
<p><a href="../2024/07/mastering-kraken2-performance-optimisation.html">Part 2 - Classification Performance Optimisation</a></p>
<p><a href="../2024/07/mastering-kraken2-build-custom-db.html">Part 3 - Build custom database indices</a> (this post)</p>
<p><a href="../2024/08/mastering-kraken2-fda-argos-index.html">Part 4 - Build FDA-ARGOS index</a> </p>
<p>Part 5 - Regular vs Fast Builds (upcoming)</p>
<p>Part 6 - Benchmarking (upcoming)</p>
<h4>Introduction</h4>
<p>In the previous post, we learned how to improve kraken2<sup id="fnref:k2"><a class="footnote-ref" href="../2024/07/mastering-kraken2-build-custom-db.html#fn:k2">1</a></sup> classification performance. So far we have downloaded &amp; used pre-built genome indices(databases). </p>
<p>In this post, let's build a custom database for kraken2. For simplicity, let's use only refseq archaea genomes<sup id="fnref:rag"><a class="footnote-ref" href="../2024/07/mastering-kraken2-build-custom-db.html#fn:rag">2</a></sup> for building the index.</p>
<h4>Building Custom Database</h4>
<p>First, we need to download the taxonomy files. We can use the <code>k2</code> script provided by kraken2.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>k2<span class="w"> </span>download-taxonomy<span class="w"> </span>--db<span class="w"> </span>custom_db
</pre></div>

<p>This takes ~30 minutes depending on the network speed. The taxonomy files are downloaded to the <code>custom_db/taxonomy</code> directory.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>ls<span class="w"> </span>custom_db/taxonomy
citations.dmp<span class="w">  </span>division.dmp<span class="w">  </span>gencode.dmp<span class="w">  </span>merged.dmp<span class="w">  </span>nodes.dmp
nucl_wgs.accession2taxid<span class="w"> </span>delnodes.dmp<span class="w">  </span>gc.prt<span class="w"> </span>
images.dmp<span class="w">  </span>names.dmp<span class="w">  </span>nucl_gb.accession2taxid<span class="w">  </span>readme.txt

$<span class="w"> </span>du<span class="w"> </span>-hs<span class="w"> </span>custom_db/taxonomy
43G<span class="w">     </span>custom_db/taxonomy
</pre></div>

<p>For simplicity, let's use the archaea refseq genomes. We can use <code>kraken2-build</code> to download the refseq genomes.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>k2<span class="w"> </span>download-library<span class="w"> </span>--library<span class="w"> </span>archaea<span class="w"> </span>--db<span class="w"> </span>custom_db
</pre></div>

<p>This runs on a single thread. Instead of using <code>kraken2-build</code>, we can use <code>ncbi-genome-download</code><sup id="fnref:ngd"><a class="footnote-ref" href="../2024/07/mastering-kraken2-build-custom-db.html#fn:ngd">3</a></sup> tool to download the genomes. This provides much granular control over the download process. For example, we can download only <code>--assembly-levels complete</code> genomes. We can also download multiple genomes in parallel.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>pip<span class="w"> </span>install<span class="w"> </span>ncbi-genome-download

$<span class="w"> </span>conda<span class="w"> </span>install<span class="w"> </span>-c<span class="w"> </span>bioconda<span class="w"> </span>ncbi-genome-download

$<span class="w"> </span>ncbi-genome-download<span class="w"> </span>-s<span class="w"> </span>refseq<span class="w"> </span>-F<span class="w"> </span>fasta<span class="w"> </span>--parallel<span class="w"> </span><span class="m">40</span><span class="w"> </span>-P<span class="w"> </span>archaea
Checking<span class="w"> </span>assemblies:<span class="w"> </span><span class="m">100</span>%<span class="p">|</span>███<span class="p">|</span><span class="w"> </span><span class="m">2184</span>/2184<span class="w"> </span><span class="o">[</span><span class="m">00</span>:19&lt;<span class="m">00</span>:00,<span class="w"> </span><span class="m">111</span>.60entries/s<span class="o">]</span>
Downloading<span class="w"> </span>assemblies:<span class="w"> </span><span class="m">100</span>%<span class="p">|</span>███<span class="p">|</span><span class="w"> </span><span class="m">2184</span>/2184<span class="w">  </span><span class="o">[</span><span class="m">02</span>:04&lt;<span class="m">00</span>:00,<span class="w">  </span><span class="m">4</span>.54s/files<span class="o">]</span>
Downloading<span class="w"> </span>assemblies:<span class="w"> </span>2184files<span class="w"> </span><span class="o">[</span><span class="m">02</span>:23,<span class="w"> </span>2184files/s<span class="o">]</span>
</pre></div>

<p>In just 2 minutes, it has downloaded all the files. Lets gunzip the files.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>find<span class="w"> </span>refseq<span class="w"> </span>-name<span class="w"> </span><span class="s2">"*.gz"</span><span class="w"> </span>-print0<span class="w"> </span><span class="p">|</span><span class="w"> </span>parallel<span class="w"> </span>-0<span class="w"> </span>gunzip

$<span class="w"> </span>du<span class="w"> </span>-hs<span class="w"> </span>refseq
<span class="m">5</span>.9G<span class="w">    </span>refseq
</pre></div>

<p>Lets add all fasta genome files to the custom database</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span><span class="nb">time</span><span class="w"> </span>find<span class="w"> </span>refseq<span class="w"> </span>-name<span class="w"> </span><span class="s2">"*.fna"</span><span class="w"> </span>-exec<span class="w"> </span>kraken2-build<span class="w"> </span>--add-to-library<span class="w"> </span><span class="o">{}</span><span class="w"> </span>--db<span class="w"> </span>custom_db<span class="w"> </span><span class="se">\;</span>
<span class="m">667</span>.46s<span class="w"> </span>user<span class="w"> </span><span class="m">90</span>.78s<span class="w"> </span>system<span class="w"> </span><span class="m">106</span>%<span class="w"> </span>cpu<span class="w"> </span><span class="m">12</span>:54.80<span class="w"> </span>total
</pre></div>

<p><code>kraken2-build</code> doesn't use multiple threads for adding genomes to the database. In addition to that, it also doesn't check if the genome is already present in the database. </p>
<p>Let's use <code>k2</code> for adding genomes to the database.</p>
<div class="code"><pre class="code literal-block"><span class="nb">export</span><span class="w"> </span><span class="nv">KRAKEN_NUM_THREADS</span><span class="o">=</span><span class="m">40</span>

$<span class="w"> </span>find<span class="w"> </span>.<span class="w"> </span>-name<span class="w"> </span><span class="s2">"*.fna"</span><span class="w"> </span>-exec<span class="w"> </span>k2<span class="w"> </span>add-to-library<span class="w"> </span>--files<span class="w"> </span><span class="o">{}</span><span class="w"> </span>--db<span class="w"> </span>custom_db<span class="w"> </span><span class="se">\;</span>
<span class="m">668</span>.37s<span class="w"> </span>user<span class="w"> </span><span class="m">88</span>.44s<span class="w"> </span>system<span class="w"> </span><span class="m">159</span>%<span class="w"> </span>cpu<span class="w"> </span><span class="m">7</span>:54.40<span class="w"> </span>total
</pre></div>

<p>This took only half the time compared to <code>kraken2-build</code>.</p>
<p>Let's build the index from the library.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span><span class="nb">time</span><span class="w"> </span>kraken2-build<span class="w"> </span>--db<span class="w"> </span>custom_db<span class="w"> </span>--build<span class="w"> </span>--threads<span class="w"> </span><span class="m">36</span>
Creating<span class="w"> </span>sequence<span class="w"> </span>ID<span class="w"> </span>to<span class="w"> </span>taxonomy<span class="w"> </span>ID<span class="w"> </span>map<span class="w"> </span><span class="o">(</span>step<span class="w"> </span><span class="m">1</span><span class="o">)</span>...
Found<span class="w"> </span><span class="m">0</span>/125783<span class="w"> </span>targets,<span class="w"> </span>searched<span class="w"> </span>through<span class="w"> </span><span class="m">60000000</span><span class="w"> </span>accession<span class="w"> </span>IDs...
Found<span class="w"> </span><span class="m">59923</span>/125783<span class="w"> </span>targets,<span class="w"> </span>searched<span class="w"> </span>through<span class="w"> </span><span class="m">822105735</span><span class="w"> </span>accession<span class="w"> </span>IDs,<span class="w"> </span>search<span class="w"> </span>complete.
lookup_accession_numbers:<span class="w"> </span><span class="m">65860</span>/125783<span class="w"> </span>accession<span class="w"> </span>numbers<span class="w"> </span>remain<span class="w"> </span>unmapped,<span class="w"> </span>see<span class="w"> </span>unmapped.txt<span class="w"> </span><span class="k">in</span><span class="w"> </span>DB<span class="w"> </span>directory
Sequence<span class="w"> </span>ID<span class="w"> </span>to<span class="w"> </span>taxonomy<span class="w"> </span>ID<span class="w"> </span>map<span class="w"> </span>complete.<span class="w"> </span><span class="o">[</span>2m1.950s<span class="o">]</span>
Estimating<span class="w"> </span>required<span class="w"> </span>capacity<span class="w"> </span><span class="o">(</span>step<span class="w"> </span><span class="m">2</span><span class="o">)</span>...
Estimated<span class="w"> </span><span class="nb">hash</span><span class="w"> </span>table<span class="w"> </span>requirement:<span class="w"> </span><span class="m">5340021028</span><span class="w"> </span>bytes
Capacity<span class="w"> </span>estimation<span class="w"> </span>complete.<span class="w"> </span><span class="o">[</span><span class="m">23</span>.875s<span class="o">]</span>
Building<span class="w"> </span>database<span class="w"> </span>files<span class="w"> </span><span class="o">(</span>step<span class="w"> </span><span class="m">3</span><span class="o">)</span>...
Taxonomy<span class="w"> </span>parsed<span class="w"> </span>and<span class="w"> </span>converted.
CHT<span class="w"> </span>created<span class="w"> </span>with<span class="w"> </span><span class="m">11</span><span class="w"> </span>bits<span class="w"> </span>reserved<span class="w"> </span><span class="k">for</span><span class="w"> </span>taxid.
Completed<span class="w"> </span>processing<span class="w"> </span>of<span class="w"> </span><span class="m">59911</span><span class="w"> </span>sequences,<span class="w"> </span><span class="m">3572145823</span><span class="w"> </span>bp
Writing<span class="w"> </span>data<span class="w"> </span>to<span class="w"> </span>disk...<span class="w">  </span>complete.
Database<span class="w"> </span>files<span class="w"> </span>completed.<span class="w"> </span><span class="o">[</span>12m3.368s<span class="o">]</span>
Database<span class="w"> </span>construction<span class="w"> </span>complete.<span class="w"> </span><span class="o">[</span>Total:<span class="w"> </span>14m29.666s<span class="o">]</span>
kraken2-build<span class="w"> </span>--db<span class="w"> </span>custom_db<span class="w"> </span>--build<span class="w"> </span>--threads<span class="w"> </span><span class="m">36</span><span class="w">  </span><span class="m">24534</span>.98s<span class="w"> </span>user<span class="w"> </span><span class="m">90</span>.50s<span class="w"> </span>system<span class="w"> </span><span class="m">2831</span>%<span class="w"> </span>cpu<span class="w"> </span><span class="m">14</span>:29.75<span class="w"> </span>total

$<span class="w"> </span>ls<span class="w"> </span>-ll
.rw-rw-r--<span class="w"> </span><span class="m">5</span>.3G<span class="w"> </span>anand<span class="w">  </span><span class="m">1</span><span class="w"> </span>Aug<span class="w"> </span><span class="m">16</span>:35<span class="w"> </span>hash.k2d
drwxrwxr-x<span class="w">    </span>-<span class="w"> </span>anand<span class="w">  </span><span class="m">1</span><span class="w"> </span>Aug<span class="w"> </span><span class="m">12</span>:32<span class="w"> </span>library
.rw-rw-r--<span class="w">   </span><span class="m">64</span><span class="w"> </span>anand<span class="w">  </span><span class="m">1</span><span class="w"> </span>Aug<span class="w"> </span><span class="m">16</span>:35<span class="w"> </span>opts.k2d
.rw-rw-r--<span class="w"> </span><span class="m">1</span>.5M<span class="w"> </span>anand<span class="w">  </span><span class="m">1</span><span class="w"> </span>Aug<span class="w"> </span><span class="m">16</span>:22<span class="w"> </span>seqid2taxid.map
.rw-rw-r--<span class="w"> </span>115k<span class="w"> </span>anand<span class="w">  </span><span class="m">1</span><span class="w"> </span>Aug<span class="w"> </span><span class="m">16</span>:23<span class="w"> </span>taxo.k2d
lrwxrwxrwx<span class="w">   </span><span class="m">20</span><span class="w"> </span>anand<span class="w">  </span><span class="m">1</span><span class="w"> </span>Aug<span class="w"> </span><span class="m">12</span>:31<span class="w"> </span>taxonomy
.rw-rw-r--<span class="w"> </span><span class="m">1</span>.2M<span class="w"> </span>anand<span class="w">  </span><span class="m">1</span><span class="w"> </span>Aug<span class="w"> </span><span class="m">16</span>:22<span class="w"> </span>unmapped.txt
</pre></div>

<p>We are able to build index for ~6GB input files in ~15 minutes.</p>
<h4>Conclusion</h4>
<p>We learnt some useful tips to speed up the custom database creation process. In the next post, we will learn about regular vs. fast builds.</p>
<div class="footnote">
<hr>
<ol>
<li id="fn:k2">
<p><a href="https://ccb.jhu.edu/software/kraken2/">Kraken2</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-build-custom-db.html#fnref:k2" title="Jump back to footnote 1 in the text">↩</a></p>
</li>
<li id="fn:rag">
<p><a href="https://ftp.ncbi.nlm.nih.gov/genomes/refseq/archaea/">RefSeq Archaea genomes</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-build-custom-db.html#fnref:rag" title="Jump back to footnote 2 in the text">↩</a></p>
</li>
<li id="fn:ngd">
<p><a href="https://github.com/kblin/ncbi-genome-download">https://github.com/kblin/ncbi-genome-download</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-build-custom-db.html#fnref:ngd" title="Jump back to footnote 3 in the text">↩</a></p>
</li>
</ol>
</div>
    </div>
    </article><article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="../2024/07/mastering-kraken2-performance-optimisation.html" class="u-url">Mastering Kraken2 - Part 2 - Performance Optimisation</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                Chillar Anand
            </span></p>
            <p class="dateline">
            <a href="../2024/07/mastering-kraken2-performance-optimisation.html" rel="bookmark">
            <time class="published dt-published" datetime="2024-07-28T10:51:30+05:30" itemprop="datePublished" title="2024-07-28">2024-07-28</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <h4>Mastering Kraken2</h4>
<p><a href="../2024/07/mastering-kraken2-initial-runs.html">Part 1 - Initial Runs</a></p>
<p><a href="../2024/07/mastering-kraken2-performance-optimisation.html">Part 2 - Classification Performance Optimisation</a> (this post)</p>
<p><a href="../2024/07/mastering-kraken2-build-custom-db.html">Part 3 - Build custom database indices</a></p>
<p><a href="../2024/08/mastering-kraken2-fda-argos-index.html">Part 4 - Build FDA-ARGOS index</a> </p>
<p>Part 5 - Regular vs Fast Builds (upcoming)</p>
<p>Part 6 - Benchmarking (upcoming)</p>
<h4>Introduction</h4>
<p>In the previous post, we learned how to set up kraken2<sup id="fnref:k2"><a class="footnote-ref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fn:k2">1</a></sup>, download pre-built indices, and run kraken2. In this post, we will learn various ways to speed up the classification process.</p>
<h4>Increasing RAM</h4>
<p>Kraken2 standard database is ~80GB in size. It is recommended to have at least db size RAM to run kraken2 efficiently<sup id="fnref:ksr"><a class="footnote-ref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fn:ksr">2</a></sup>. Let's use 128GB RAM machine and run kraken2 with ERR10359977<sup id="fnref:err"><a class="footnote-ref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fn:err">3</a></sup> sample.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span><span class="nb">time</span><span class="w"> </span>kraken2<span class="w"> </span>--db<span class="w"> </span>k2_standard<span class="w"> </span>--report<span class="w"> </span>report.txt<span class="w"> </span>ERR10359977.fastq.gz<span class="w"> </span>&gt;<span class="w"> </span>output.txt
Loading<span class="w"> </span>database<span class="w"> </span>information...<span class="w"> </span><span class="k">done</span>.
<span class="m">95064</span><span class="w"> </span>sequences<span class="w"> </span><span class="o">(</span><span class="m">14</span>.35<span class="w"> </span>Mbp<span class="o">)</span><span class="w"> </span>processed<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">2</span>.142s<span class="w"> </span><span class="o">(</span><span class="m">2662</span>.9<span class="w"> </span>Kseq/m,<span class="w"> </span><span class="m">402</span>.02<span class="w"> </span>Mbp/m<span class="o">)</span>.
<span class="w">  </span><span class="m">94816</span><span class="w"> </span>sequences<span class="w"> </span>classified<span class="w"> </span><span class="o">(</span><span class="m">99</span>.74%<span class="o">)</span>
<span class="w">  </span><span class="m">248</span><span class="w"> </span>sequences<span class="w"> </span>unclassified<span class="w"> </span><span class="o">(</span><span class="m">0</span>.26%<span class="o">)</span>
kraken2<span class="w"> </span>--db<span class="w"> </span>k2_standard<span class="w"> </span>--report<span class="w"> </span>report.txt<span class="w"> </span>ERR10359977.fastq.gz<span class="w"> </span>&gt;<span class="w">   </span><span class="m">1</span>.68s<span class="w"> </span>user<span class="w"> </span><span class="m">152</span>.19s<span class="w"> </span>system<span class="w"> </span><span class="m">35</span>%<span class="w"> </span>cpu<span class="w"> </span><span class="m">7</span>:17.55<span class="w"> </span>total
</pre></div>

<p>Now the time taken has come down from 40 minutes to 7 minutes. The classification speed has also increased from 0.19 Mbp/m to 402.02 Mbp/m.</p>
<p>The previous sample had only a few reads, and the speed is not a good indicator. Let's run kraken2 with a larger sample.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span><span class="nb">time</span><span class="w"> </span>kraken2<span class="w"> </span>--db<span class="w"> </span>k2_standard<span class="w"> </span>--report<span class="w"> </span>report.txt<span class="w"> </span>--paired<span class="w"> </span>SRR6915097_1.fastq.gz<span class="w"> </span>SRR6915097_2.fastq.gz<span class="w"> </span>&gt;<span class="w"> </span>output.txt
Loading<span class="w"> </span>database<span class="w"> </span>information...<span class="w"> </span><span class="k">done</span>.
Processed<span class="w"> </span><span class="m">14980000</span><span class="w"> </span>sequences<span class="w"> </span><span class="o">(</span><span class="m">2972330207</span><span class="w"> </span>bp<span class="o">)</span><span class="w"> </span>...
<span class="m">17121245</span><span class="w"> </span>sequences<span class="w"> </span><span class="o">(</span><span class="m">3397</span>.15<span class="w"> </span>Mbp<span class="o">)</span><span class="w"> </span>processed<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">797</span>.424s<span class="w"> </span><span class="o">(</span><span class="m">1288</span>.2<span class="w"> </span>Kseq/m,<span class="w"> </span><span class="m">255</span>.61<span class="w"> </span>Mbp/m<span class="o">)</span>.
<span class="w">  </span><span class="m">9826671</span><span class="w"> </span>sequences<span class="w"> </span>classified<span class="w"> </span><span class="o">(</span><span class="m">57</span>.39%<span class="o">)</span>
<span class="w">  </span><span class="m">7294574</span><span class="w"> </span>sequences<span class="w"> </span>unclassified<span class="w"> </span><span class="o">(</span><span class="m">42</span>.61%<span class="o">)</span>
kraken2<span class="w"> </span>--db<span class="w"> </span>k2_standard<span class="w"> </span>--report<span class="w"> </span>report.txt<span class="w"> </span>--paired<span class="w"> </span>&gt;<span class="w"> </span>output.txt<span class="w">  </span><span class="m">526</span>.39s<span class="w"> </span>user<span class="w"> </span><span class="m">308</span>.24s<span class="w"> </span>system<span class="w"> </span><span class="m">68</span>%<span class="w"> </span>cpu<span class="w"> </span><span class="m">20</span>:23.86<span class="w"> </span>total
</pre></div>

<p>This took almost 20 minutes to classify ~3 Gbp of data. Out of 20 minutes, 13 minutes was spent in classification. The remaining time in loading the db into memory.</p>
<p>Let's use k2_plusPF<sup id="fnref:k2p"><a class="footnote-ref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fn:k2p">4</a></sup> db, which is twice the size of k2_standard and run kraken2.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span><span class="nb">time</span><span class="w"> </span>kraken2<span class="w"> </span>--db<span class="w"> </span>k2_plusfp<span class="w"> </span>--report<span class="w"> </span>report.txt<span class="w"> </span>--paired<span class="w"> </span>SRR6915097_1.fastq.gz<span class="w"> </span>SRR6915097_2.fastq.gz<span class="w"> </span>&gt;<span class="w"> </span>output.txt
Loading<span class="w"> </span>database<span class="w"> </span>information...done.
<span class="m">17121245</span><span class="w"> </span>sequences<span class="w"> </span><span class="o">(</span><span class="m">3397</span>.15<span class="w"> </span>Mbp<span class="o">)</span><span class="w"> </span>processed<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">755</span>.290s<span class="w"> </span><span class="o">(</span><span class="m">1360</span>.1<span class="w"> </span>Kseq/m,<span class="w"> </span><span class="m">269</span>.87<span class="w"> </span>Mbp/m<span class="o">)</span>.
<span class="w">  </span><span class="m">9903824</span><span class="w"> </span>sequences<span class="w"> </span>classified<span class="w"> </span><span class="o">(</span><span class="m">57</span>.85%<span class="o">)</span>
<span class="w">  </span><span class="m">7217421</span><span class="w"> </span>sequences<span class="w"> </span>unclassified<span class="w"> </span><span class="o">(</span><span class="m">42</span>.15%<span class="o">)</span>
kraken2<span class="w"> </span>--db<span class="w"> </span>k2_plusfp/<span class="w"> </span>--report<span class="w"> </span>report.txt<span class="w"> </span>--paired<span class="w"> </span>SRR6915097_1.fastq.gz<span class="w">  </span>&gt;<span class="w">   </span><span class="m">509</span>.71s<span class="w"> </span>user<span class="w"> </span><span class="m">509</span>.51s<span class="w"> </span>system<span class="w"> </span><span class="m">55</span>%<span class="w"> </span>cpu<span class="w"> </span><span class="m">30</span>:35.49<span class="w"> </span>total
</pre></div>

<p>This took ~30 minutes to complete, but the classification took only 13 minutes similar to k2_standard. The remaining time was spent in loading the db into memory.</p>
<h4>Preloading db into RAM</h4>
<p>We can use vmtouch<sup id="fnref:vmt"><a class="footnote-ref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fn:vmt">5</a></sup> to preload db into RAM. kraken2 provides <code>--memory-mapping</code> option to use preloaded db. </p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>vmtouch<span class="w"> </span>-vt<span class="w"> </span>k2_standard/hash.k2d<span class="w"> </span>k2_standard/opts.k2d<span class="w"> </span>k2_standard/taxo.k2d
<span class="w">           </span>Files:<span class="w"> </span><span class="m">3</span>
<span class="w">     </span>Directories:<span class="w"> </span><span class="m">0</span>
<span class="w">   </span>Touched<span class="w"> </span>Pages:<span class="w"> </span><span class="m">20382075</span><span class="w"> </span><span class="o">(</span>77G<span class="o">)</span>
<span class="w">         </span>Elapsed:<span class="w"> </span><span class="m">434</span>.77<span class="w"> </span>seconds
</pre></div>

<p>When Linux requires RAM, it will incrementally evict the db from memory. To prevent this, we can copy the db to shared memory (/dev/shm) and then use vmtouch to preload the db.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>cp<span class="w"> </span>-r<span class="w"> </span>k2_standard<span class="w"> </span>/dev/shm

$<span class="w"> </span>vmtouch<span class="w"> </span>-t<span class="w"> </span>/dev/shm/*.k2d
</pre></div>

<p>Now, let's run kraken2 with <code>--memory-mapping</code> option.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span><span class="nb">time</span><span class="w"> </span>kraken2<span class="w"> </span>--db<span class="w"> </span>k2_standard<span class="w"> </span>--report<span class="w"> </span>report.txt<span class="w"> </span>--memory-mapping<span class="w"> </span>--paired<span class="w"> </span>SRR6915097_1.fastq.gz<span class="w"> </span>SRR6915097_2.fastq.gz<span class="w"> </span>&gt;<span class="w"> </span>output.txt
Loading<span class="w"> </span>database<span class="w"> </span>information...<span class="w"> </span><span class="k">done</span>.
<span class="m">17121245</span><span class="w"> </span>sequences<span class="w"> </span><span class="o">(</span><span class="m">3397</span>.15<span class="w"> </span>Mbp<span class="o">)</span><span class="w"> </span>processed<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">532</span>.486s<span class="w"> </span><span class="o">(</span><span class="m">1929</span>.2<span class="w"> </span>Kseq/m,<span class="w"> </span><span class="m">382</span>.79<span class="w"> </span>Mbp/m<span class="o">)</span>.
<span class="w">  </span><span class="m">9826671</span><span class="w"> </span>sequences<span class="w"> </span>classified<span class="w"> </span><span class="o">(</span><span class="m">57</span>.39%<span class="o">)</span>
<span class="w">  </span><span class="m">7294574</span><span class="w"> </span>sequences<span class="w"> </span>unclassified<span class="w"> </span><span class="o">(</span><span class="m">42</span>.61%<span class="o">)</span>
<span class="w">  </span>kraken2<span class="w"> </span>--db<span class="w"> </span>k2_standard<span class="w"> </span>--report<span class="w"> </span>report.txt<span class="w"> </span>--paired<span class="w"> </span>SRR6915097_1.fastq.gz<span class="w">   </span>&gt;<span class="w">  </span><span class="m">424</span>.20s<span class="w"> </span>user<span class="w"> </span><span class="m">11</span>.76s<span class="w"> </span>system<span class="w"> </span><span class="m">81</span>%<span class="w"> </span>cpu<span class="w"> </span><span class="m">8</span>:54.98<span class="w"> </span>total
</pre></div>

<p>Now the classification took only ~10 minutes.</p>
<h4>Multi threading</h4>
<p>kraken2 supports multiple threads. I am using a machine with 40 threads.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span><span class="nb">time</span><span class="w"> </span>kraken2<span class="w"> </span>--db<span class="w"> </span>k2_standard<span class="w"> </span>--report<span class="w"> </span>report.txt<span class="w"> </span>--paired<span class="w"> </span>SRR6915097_1.fastq.gz<span class="w"> </span>SRR6915097_2.fastq.gz<span class="w"> </span>--memory-mapping<span class="w"> </span>--threads<span class="w"> </span><span class="m">32</span><span class="w"> </span>&gt;<span class="w"> </span>output.txt
Loading<span class="w"> </span>database<span class="w"> </span>information...<span class="w"> </span><span class="k">done</span>.
<span class="m">17121245</span><span class="w"> </span>sequences<span class="w"> </span><span class="o">(</span><span class="m">3397</span>.15<span class="w"> </span>Mbp<span class="o">)</span><span class="w"> </span>processed<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">71</span>.675s<span class="w"> </span><span class="o">(</span><span class="m">14332</span>.5<span class="w"> </span>Kseq/m,<span class="w"> </span><span class="m">2843</span>.81<span class="w"> </span>Mbp/m<span class="o">)</span>.
<span class="w">  </span><span class="m">9826671</span><span class="w"> </span>sequences<span class="w"> </span>classified<span class="w"> </span><span class="o">(</span><span class="m">57</span>.39%<span class="o">)</span>
<span class="w">  </span><span class="m">7294574</span><span class="w"> </span>sequences<span class="w"> </span>unclassified<span class="w"> </span><span class="o">(</span><span class="m">42</span>.61%<span class="o">)</span>
kraken2<span class="w"> </span>--db<span class="w"> </span>k2_standard<span class="w"> </span>--report<span class="w"> </span>report.txt<span class="w"> </span>--paired<span class="w"> </span>SRR6915097_1.fastq.gz<span class="w">      </span><span class="m">556</span>.58s<span class="w"> </span>user<span class="w"> </span><span class="m">22</span>.85s<span class="w"> </span>system<span class="w"> </span><span class="m">762</span>%<span class="w"> </span>cpu<span class="w"> </span><span class="m">1</span>:16.02<span class="w"> </span>total
</pre></div>

<p>With 32 threads, the classification took only 1 minute. Beyond 32 threads, the classification time did not decrease significantly.</p>
<h4>Optimising input files</h4>
<p>So far we have used gzipped input files. Let's use unzipped input files and run kraken2.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>gunzip<span class="w"> </span>SRR6915097_1.fastq.gz
$<span class="w"> </span>gunzip<span class="w"> </span>SRR6915097_2.fastq.gz

$<span class="w"> </span><span class="nb">time</span><span class="w"> </span>kraken2<span class="w"> </span>--db<span class="w"> </span>k2_standard<span class="w"> </span>--report<span class="w"> </span>report.txt<span class="w"> </span>--paired<span class="w"> </span>SRR6915097_1.fastq<span class="w"> </span>SRR6915097_2.fastq<span class="w"> </span>--memory-mapping<span class="w"> </span>--threads<span class="w"> </span><span class="m">30</span><span class="w"> </span>&gt;<span class="w"> </span>output.txt
Loading<span class="w"> </span>database<span class="w"> </span>information...<span class="w"> </span><span class="k">done</span>.
<span class="m">17121245</span><span class="w"> </span>sequences<span class="w"> </span><span class="o">(</span><span class="m">3397</span>.15<span class="w"> </span>Mbp<span class="o">)</span><span class="w"> </span>processed<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">34</span>.809s<span class="w"> </span><span class="o">(</span><span class="m">29512</span>.0<span class="w"> </span>Kseq/m,<span class="w"> </span><span class="m">5855</span>.68<span class="w"> </span>Mbp/m<span class="o">)</span>.
<span class="w">  </span><span class="m">9826671</span><span class="w"> </span>sequences<span class="w"> </span>classified<span class="w"> </span><span class="o">(</span><span class="m">57</span>.39%<span class="o">)</span>
<span class="w">  </span><span class="m">7294574</span><span class="w"> </span>sequences<span class="w"> </span>unclassified<span class="w"> </span><span class="o">(</span><span class="m">42</span>.61%<span class="o">)</span>
kraken2<span class="w"> </span>--db<span class="w"> </span>k2_standard<span class="w"> </span>--report<span class="w"> </span>report.txt<span class="w"> </span>--paired<span class="w"> </span>SRR6915097_1.fastq<span class="w">    </span><span class="m">30</span><span class="w">   </span><span class="m">565</span>.03s<span class="w"> </span>user<span class="w"> </span><span class="m">17</span>.12s<span class="w"> </span>system<span class="w"> </span><span class="m">1530</span>%<span class="w"> </span>cpu<span class="w"> </span><span class="m">38</span>.047<span class="w"> </span>total
</pre></div>

<p>Now the classification time has come down to 40 seconds.</p>
<p>Since the input fastq files are paired, interleaving the files also takes time. Lets interleave the files and run kraken2.</p>
<p>To interleave the files, lets use <code>seqfu</code> tool.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>conda<span class="w"> </span>install<span class="w"> </span>-y<span class="w"> </span>-c<span class="w"> </span>conda-forge<span class="w"> </span>-c<span class="w"> </span>bioconda<span class="w"> </span><span class="s2">"seqfu&gt;1.10"</span>

$<span class="w"> </span>seqfu<span class="w"> </span>interleave<span class="w"> </span>-1<span class="w"> </span>SRR6915097_1.fastq.gz<span class="w"> </span>-2<span class="w"> </span>SRR6915097_2.fastq.gz<span class="w"> </span>&gt;<span class="w"> </span>SRR6915097.fastq

$<span class="w"> </span><span class="nb">time</span><span class="w"> </span>kraken2<span class="w"> </span>--db<span class="w"> </span>k2_standard<span class="w"> </span>--report<span class="w"> </span>report.txt<span class="w"> </span>--memory-mapping<span class="w"> </span>SRR6915097.fq<span class="w"> </span>--threads<span class="w"> </span><span class="m">32</span><span class="w"> </span>&gt;<span class="w"> </span>output.txt
Loading<span class="w"> </span>database<span class="w"> </span>information...<span class="w"> </span><span class="k">done</span>.
<span class="m">34242490</span><span class="w"> </span>sequences<span class="w"> </span><span class="o">(</span><span class="m">3397</span>.15<span class="w"> </span>Mbp<span class="o">)</span><span class="w"> </span>processed<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">20</span>.199s<span class="w"> </span><span class="o">(</span><span class="m">101714</span>.1<span class="w"> </span>Kseq/m,<span class="w"> </span><span class="m">10090</span>.91<span class="w"> </span>Mbp/m<span class="o">)</span>.
<span class="w">  </span><span class="m">17983321</span><span class="w"> </span>sequences<span class="w"> </span>classified<span class="w"> </span><span class="o">(</span><span class="m">52</span>.52%<span class="o">)</span>
<span class="w">  </span><span class="m">16259169</span><span class="w"> </span>sequences<span class="w"> </span>unclassified<span class="w"> </span><span class="o">(</span><span class="m">47</span>.48%<span class="o">)</span>
kraken2<span class="w"> </span>--db<span class="w"> </span>k2_standard<span class="w"> </span>--report<span class="w"> </span>report.txt<span class="w"> </span>--memory-mapping<span class="w"> </span>SRR6915097.fq<span class="w">  </span><span class="m">32</span><span class="w">  </span><span class="m">618</span>.96s<span class="w"> </span>user<span class="w"> </span><span class="m">18</span>.24s<span class="w"> </span>system<span class="w"> </span><span class="m">2653</span>%<span class="w"> </span>cpu<span class="w"> </span><span class="m">24</span>.013<span class="w"> </span>total
</pre></div>

<p>Now the classification time has come down to 24 seconds. </p>
<h4>Conclusion</h4>
<p>In terms of classification speed, we have come a long way from 0.1 Mbp/m to 1200 Mbp/m. In the next post, we will learn how to optimise the creation of custom indices.</p>
<div class="footnote">
<hr>
<ol>
<li id="fn:k2">
<p><a href="https://ccb.jhu.edu/software/kraken2/">Kraken2</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fnref:k2" title="Jump back to footnote 1 in the text">↩</a></p>
</li>
<li id="fn:ksr">
<p><a href="https://github.com/DerrickWood/kraken2/blob/master/docs/MANUAL.markdown#system-requirements">Kraken System Requirements</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fnref:ksr" title="Jump back to footnote 2 in the text">↩</a></p>
</li>
<li id="fn:err">
<p><a href="ftp://ftp.sra.ebi.ac.uk/vol1/fastq/ERR103/077/ERR10359977/ERR10359977.fastq.gz">ERR10359977.fastq.gz</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fnref:err" title="Jump back to footnote 3 in the text">↩</a></p>
</li>
<li id="fn:k2p">
<p><a href="https://benlangmead.github.io/aws-indexes/k2">Genomic Index Zone - k2</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fnref:k2p" title="Jump back to footnote 4 in the text">↩</a></p>
</li>
<li id="fn:vmt">
<p><a href="https://hoytech.com/vmtouch/">https://hoytech.com/vmtouch/</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-performance-optimisation.html#fnref:vmt" title="Jump back to footnote 5 in the text">↩</a></p>
</li>
</ol>
</div>
    </div>
    </article><article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="../2024/07/mastering-kraken2-initial-runs.html" class="u-url">Mastering Kraken2 - Part 1 - Initial Runs</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                Chillar Anand
            </span></p>
            <p class="dateline">
            <a href="../2024/07/mastering-kraken2-initial-runs.html" rel="bookmark">
            <time class="published dt-published" datetime="2024-07-28T10:44:25+05:30" itemprop="datePublished" title="2024-07-28">2024-07-28</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <h4>Mastering Kraken2</h4>
<p><a href="../2024/07/mastering-kraken2-initial-runs.html">Part 1 - Initial Runs</a> (this post)</p>
<p><a href="../2024/07/mastering-kraken2-performance-optimisation.html">Part 2 - Classification Performance Optimisation</a></p>
<p><a href="../2024/07/mastering-kraken2-build-custom-db.html">Part 3 - Build custom database indices</a></p>
<p><a href="../2024/08/mastering-kraken2-fda-argos-index.html">Part 4 - Build FDA-ARGOS index</a> (this post)</p>
<p>Part 5 - Regular vs Fast Builds (upcoming)</p>
<p>Part 6 - Benchmarking (upcoming)</p>
<h4>Introduction</h4>
<p>Kraken2<sup id="fnref:Kraken2"><a class="footnote-ref" href="../2024/07/mastering-kraken2-initial-runs.html#fn:Kraken2">1</a></sup> is widely used for metagenomics taxonomic classification, and it has pre-built indexes for many organisms. In this series, we will learn</p>
<ul>
<li>How to set up kraken2, download pre-built indices</li>
<li>Run kraken2 (8GB RAM) at ~0.19 Mbp/m (million base pairs per minute)</li>
<li>Learn various ways to speed up the classification process</li>
<li>Run kraken2 (128GB RAM) at ~1200 Mbp/m</li>
<li>Build custom indices</li>
</ul>
<h4>Installation</h4>
<p>We can install kraken2 from source using the <code>install_kraken2.sh</code> script as per the manual<sup id="fnref:install_kraken2"><a class="footnote-ref" href="../2024/07/mastering-kraken2-initial-runs.html#fn:install_kraken2">2</a></sup>.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>git<span class="w"> </span>clone<span class="w"> </span>https://github.com/DerrickWood/kraken2
$<span class="w"> </span><span class="nb">cd</span><span class="w"> </span>kraken2
$<span class="w"> </span>./install_kraken2.sh<span class="w"> </span>/usr/local/bin
<span class="c1"># ensure kraken2 is in the PATH</span>
$<span class="w"> </span><span class="nb">export</span><span class="w"> </span><span class="nv">PATH</span><span class="o">=</span><span class="nv">$PATH</span>:/usr/local/bin
</pre></div>

<p>If you already have conda installed, you can install kraken2 from conda as well.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>conda<span class="w"> </span>install<span class="w"> </span>-c<span class="w"> </span>bioconda<span class="w"> </span>kraken2
</pre></div>

<p>If you have <code>brew</code> installed on Linux or Mac(including M1), you can install kraken2 using <code>brew</code>.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>brew<span class="w"> </span>install<span class="w"> </span>brewsci/bio/kraken2
</pre></div>

<h4>Download pre-built indices</h4>
<p>Building kraken2 indices take a lot of time and resources. For now, let's download and use the pre-built indices. In the final post, we will learn how to build the indices.</p>
<p>Genomic Index Zone<sup id="fnref:GenomicIndexZone"><a class="footnote-ref" href="../2024/07/mastering-kraken2-initial-runs.html#fn:GenomicIndexZone">3</a></sup> provides pre-built indices for kraken2. Let's download the standard database. It contains Refeq archaea, bacteria, viral, plasmid, human1, &amp; UniVec_Core. </p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>wget<span class="w"> </span>https://genome-idx.s3.amazonaws.com/kraken/k2_standard_20240605.tar.gz
$<span class="w"> </span>mkdir<span class="w"> </span>k2_standard
$<span class="w"> </span>tar<span class="w"> </span>-xvf<span class="w"> </span>k2_standard_20240605.tar.gz<span class="w"> </span>-C<span class="w"> </span>k2_standard
</pre></div>

<p>The extracted directory contains three files - <code>hash.k2d</code>, <code>opts.k2d</code>, <code>taxo.k2d</code> which are the kraken2 database files.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>ls<span class="w"> </span>-l<span class="w"> </span>*.k2d
.rw-r--r--<span class="w">  </span>83G<span class="w"> </span>anand<span class="w"> </span><span class="m">13</span><span class="w"> </span>Jul<span class="w"> </span><span class="m">12</span>:34<span class="w"> </span>hash.k2d
.rw-r--r--<span class="w">   </span><span class="m">64</span><span class="w"> </span>anand<span class="w"> </span><span class="m">13</span><span class="w"> </span>Jul<span class="w"> </span><span class="m">12</span>:34<span class="w"> </span>opts.k2d
.rw-r--r--<span class="w"> </span><span class="m">4</span>.0M<span class="w"> </span>anand<span class="w"> </span><span class="m">13</span><span class="w"> </span>Jul<span class="w"> </span><span class="m">12</span>:34<span class="w"> </span>taxo.k2d
</pre></div>

<h4>Classification</h4>
<p>To run the taxonomic classification, let's use <code>ERR10359977</code> human gut meta genome from NCBI SRA.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>wget<span class="w"> </span>https://ftp.sra.ebi.ac.uk/vol1/fastq/ERR103/077/ERR10359977/ERR10359977.fastq.gz
$<span class="w"> </span>kraken2<span class="w"> </span>--db<span class="w"> </span>k2_standard<span class="w"> </span>--report<span class="w"> </span>report.txt<span class="w"> </span>ERR10359977.fastq.gz<span class="w"> </span>&gt;<span class="w"> </span>output.txt
</pre></div>

<p>By default, the machine I have used has 8GB RAM and an additioinal 8GB swap. Since kraken2 needs entire db(~80GB) in memory, when the process tries to consume more than 16GB memory, the kernel will kill the process. </p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span><span class="nb">time</span><span class="w"> </span>kraken2<span class="w"> </span>--db<span class="w"> </span>k2_standard<span class="w"> </span>--paired<span class="w"> </span>SRR6915097_1.fastq.gz<span class="w"> </span>SRR6915097_2.fastq.gz<span class="w"> </span>&gt;<span class="w"> </span>output.txt
Loading<span class="w"> </span>database<span class="w"> </span>information...Command<span class="w"> </span>terminated<span class="w"> </span>by<span class="w"> </span>signal<span class="w"> </span><span class="m">9</span>
<span class="m">0</span>.02user<span class="w"> </span><span class="m">275</span>.83system<span class="w"> </span><span class="m">8</span>:17.43elapsed<span class="w"> </span><span class="m">55</span>%CPU<span class="w"> </span>
</pre></div>

<p>To prevent this, let's increase the swap space to 128 GB.</p>
<div class="code"><pre class="code literal-block"><span class="c1"># Create an empty swapfile of 128GB</span>
sudo<span class="w"> </span>dd<span class="w"> </span><span class="k">if</span><span class="o">=</span>/dev/zero<span class="w"> </span><span class="nv">of</span><span class="o">=</span>/swapfile<span class="w"> </span><span class="nv">bs</span><span class="o">=</span>1G<span class="w"> </span><span class="nv">count</span><span class="o">=</span><span class="m">128</span>

<span class="c1"># Turn swap off - It might take several minutes</span>
sudo<span class="w"> </span>swapoff<span class="w"> </span>-a

<span class="c1"># Set the permissions for swapfile</span>
sudo<span class="w"> </span>chmod<span class="w"> </span><span class="m">0600</span><span class="w"> </span>/swapfile

<span class="c1"># make it a swap area</span>
sudo<span class="w"> </span>mkswap<span class="w"> </span>/swapfile<span class="w">  </span>

<span class="c1"># Turn the swap on</span>
sudo<span class="w"> </span>swapon<span class="w"> </span>/swapfile
</pre></div>

<p>We can time the classification process using the <code>time</code> command.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span><span class="nb">time</span><span class="w"> </span>kraken2<span class="w"> </span>--db<span class="w"> </span>k2_standard<span class="w"> </span>--report<span class="w"> </span>report.txt<span class="w"> </span>ERR10359977.fastq.gz<span class="w"> </span>&gt;<span class="w"> </span>output.txt
</pre></div>

<p>If you have a machine with large RAM, the same scenario can be simulated using <code>systemd-run</code>. This will limit the memory usage of kraken2 to 6.5GB. </p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span><span class="nb">time</span><span class="w"> </span>systemd-run<span class="w"> </span>--scope<span class="w"> </span>-p<span class="w"> </span><span class="nv">MemoryMax</span><span class="o">=</span><span class="m">6</span>.5G<span class="w"> </span>--user<span class="w"> </span><span class="nb">time</span><span class="w"> </span>kraken2<span class="w"> </span>--db<span class="w"> </span>k2_standard<span class="w"> </span>--report<span class="w"> </span>report.txt<span class="w"> </span>ERR10359977.fastq.gz<span class="w"> </span>&gt;<span class="w"> </span>output.txt
</pre></div>

<p>Depending on the CPU performance, this will take around ~40 minutes to complete.</p>
<div class="code"><pre class="code literal-block">Loading<span class="w"> </span>database<span class="w"> </span>information...<span class="w"> </span><span class="k">done</span>.
<span class="m">95064</span><span class="w"> </span>sequences<span class="w"> </span><span class="o">(</span><span class="m">14</span>.35<span class="w"> </span>Mbp<span class="o">)</span><span class="w"> </span>processed<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="m">1026</span>.994s<span class="w"> </span><span class="o">(</span><span class="m">5</span>.6<span class="w"> </span>Kseq/m,<span class="w"> </span><span class="m">0</span>.84<span class="w"> </span>Mbp/m<span class="o">)</span>.
<span class="w">  </span><span class="m">94939</span><span class="w"> </span>sequences<span class="w"> </span>classified<span class="w"> </span><span class="o">(</span><span class="m">99</span>.87%<span class="o">)</span>
<span class="w">  </span><span class="m">125</span><span class="w"> </span>sequences<span class="w"> </span>unclassified<span class="w"> </span><span class="o">(</span><span class="m">0</span>.13%<span class="o">)</span>
<span class="w">  </span><span class="m">4</span>.24user<span class="w"> </span><span class="m">658</span>.68system<span class="w"> </span><span class="m">38</span>:26.78elapsed<span class="w"> </span><span class="m">28</span>%CPU<span class="w"> </span>
</pre></div>

<p>If we try gut WGS(Whole Genome Sequence) sample like <code>SRR6915097</code> <sup id="fnref:srr1"><a class="footnote-ref" href="../2024/07/mastering-kraken2-initial-runs.html#fn:srr1">4</a></sup><sup id="fnref:srr2"><a class="footnote-ref" href="../2024/07/mastering-kraken2-initial-runs.html#fn:srr2">5</a></sup>. which contains ~3.3 Gbp, it will take weeks to complete.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>wget<span class="w"> </span>-c<span class="w"> </span>https://ftp.sra.ebi.ac.uk/vol1/fastq/SRR691/007/SRR6915097/SRR6915097_1.fastq.gz
$<span class="w"> </span>wget<span class="w"> </span>-c<span class="w"> </span>https://ftp.sra.ebi.ac.uk/vol1/fastq/SRR691/007/SRR6915097/SRR6915097_2.fastq.gz

$<span class="w"> </span><span class="nb">time</span><span class="w"> </span>systemd-run<span class="w"> </span>--scope<span class="w"> </span>-p<span class="w"> </span><span class="nv">MemoryMax</span><span class="o">=</span>6G<span class="w"> </span>--user<span class="w"> </span><span class="nb">time</span><span class="w"> </span>kraken2<span class="w"> </span>--db<span class="w"> </span>k2_standard<span class="w"> </span>--paired<span class="w"> </span>SRR6915097_1.fastq.gz<span class="w"> </span>SRR6915097_2.fastq.gz<span class="w"> </span>&gt;<span class="w"> </span>output.txt
</pre></div>

<p>I tried running this on 8 GB machine. Even after 10 days, it processed only 10% of the data.</p>
<p>If we have to process a large number of such samples, it takes months and this is not a practical solution. </p>
<h4>Conclusion</h4>
<p>In this post, we ran kraken2 on an 8GB machine and learned that it is not feasible to run kraken2 on large samples.</p>
<p>In the next post, we will learn how to speed up the classification process and run classification at 1200 Mbp/m.</p>
<p><strong>Next</strong>: <a href="../2024/07/mastering-kraken2-performance-optimisation.html">Part 2 - Performance Optimisation</a></p>
<div class="footnote">
<hr>
<ol>
<li id="fn:Kraken2">
<p><a href="https://ccb.jhu.edu/software/kraken2/">Kraken2</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-initial-runs.html#fnref:Kraken2" title="Jump back to footnote 1 in the text">↩</a></p>
</li>
<li id="fn:install_kraken2">
<p><a href="https://github.com/DerrickWood/kraken2/blob/master/docs/MANUAL.markdown#installation">Kraken2 - Manual - Install</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-initial-runs.html#fnref:install_kraken2" title="Jump back to footnote 2 in the text">↩</a></p>
</li>
<li id="fn:GenomicIndexZone">
<p><a href="https://benlangmead.github.io/aws-indexes/k2">Genomic Index Zone - k2</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-initial-runs.html#fnref:GenomicIndexZone" title="Jump back to footnote 3 in the text">↩</a></p>
</li>
<li id="fn:srr1">
<p><a href="https://ftp.sra.ebi.ac.uk/vol1/fastq/SRR691/007/SRR6915097/SRR6915097_1.fastq.gz">SRR6915097_1.fastq.gz</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-initial-runs.html#fnref:srr1" title="Jump back to footnote 4 in the text">↩</a></p>
</li>
<li id="fn:srr2">
<p><a href="https://ftp.sra.ebi.ac.uk/vol1/fastq/SRR691/007/SRR6915097/SRR6915097_2.fastq.gz">SRR6915097_1.fastq.gz</a> <a class="footnote-backref" href="../2024/07/mastering-kraken2-initial-runs.html#fnref:srr2" title="Jump back to footnote 5 in the text">↩</a></p>
</li>
</ol>
</div>
    </div>
    </article><article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="../2024/06/headlamp-k8s-lens-open-source-alternative.html" class="u-url">Headlamp - k8s Lens open source alternative</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                Chillar Anand
            </span></p>
            <p class="dateline">
            <a href="../2024/06/headlamp-k8s-lens-open-source-alternative.html" rel="bookmark">
            <time class="published dt-published" datetime="2024-06-24T01:48:02+05:30" itemprop="datePublished" title="2024-06-24">2024-06-24</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <p><img alt="headlamp - Open source Kubernetes Lens alternator" src="../images/headlamp-k8s-lens-open-source-alternative.png"></p>
<p>Since Lens is not open source, I tried out monokle, octant, k9s, and headlamp<sup id="fnref:headlamp"><a class="footnote-ref" href="../2024/06/headlamp-k8s-lens-open-source-alternative.html#fn:headlamp">1</a></sup>. Among them, headlamp UI &amp; features are closest to Lens. </p>
<h4>Headlamp</h4>
<p>Headlamp is CNCF sandbox project that provides cross-platform desktop application to manage Kubernetes clusters. It auto-detects clusters and provides cluster wide resource usage by default. </p>
<p>It can also be installed inside the cluster and can be accessed using a web browser. This is useful when we want to access the cluster from a mobile device.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>helm<span class="w"> </span>repo<span class="w"> </span>add<span class="w"> </span>headlamp<span class="w"> </span>https://headlamp-k8s.github.io/headlamp/

$<span class="w"> </span>helm<span class="w"> </span>install<span class="w"> </span>headlamp<span class="w"> </span>headlamp/headlamp
</pre></div>

<p>Lets port-forward the service &amp; copy the token to access it.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>kubectl<span class="w"> </span>create<span class="w"> </span>token<span class="w"> </span>headlamp

<span class="c1"># we can do this via headlamp UI as well</span>
$<span class="w"> </span>kubectl<span class="w"> </span>port-forward<span class="w"> </span>service/headlamp<span class="w"> </span><span class="m">8080</span>:80
</pre></div>

<p>Now, we can access the headlamp UI at <a href="http://">http://localhost:8080</a>.</p>
<p><img alt="headlamp - Open source Kubernetes Lens alternator" src="../images/headlamp-k8s-lens-open-source-alternative2.png"></p>
<h4>Conclusion</h4>
<p>If you are looking for an open source alternative to Lens, headlamp is a good choice. It provides a similar UI &amp; features as Lens, and it is accessible via mobile devices as well. </p>
<div class="footnote">
<hr>
<ol>
<li id="fn:headlamp">
<p><a href="https://headlamp.dev/">https://headlamp.dev/</a> <a class="footnote-backref" href="../2024/06/headlamp-k8s-lens-open-source-alternative.html#fnref:headlamp" title="Jump back to footnote 1 in the text">↩</a></p>
</li>
</ol>
</div>
    </div>
    </article><article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="../2024/06/macos-log-track-cpu-ram-usage.html" class="u-url">macOS - Log &amp; track historical CPU, RAM usage</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                Chillar Anand
            </span></p>
            <p class="dateline">
            <a href="../2024/06/macos-log-track-cpu-ram-usage.html" rel="bookmark">
            <time class="published dt-published" datetime="2024-06-01T01:48:02+05:30" itemprop="datePublished" title="2024-06-01">2024-06-01</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <p><img alt="macOS - Log CPU &amp; RAM history" src="../images/mac-log-cpu-ram-grafana.png"></p>
<p>In macOS, we can use inbuilt <code>Activity Monitor</code> or third party apps like <code>Stats</code> to check the live CPU/RAM usage. But, we can't track the historical CPU &amp; memory usage. <code>sar</code>, <code>atop</code> can track the historical CPU &amp; memory usage. But, they are not available for macOS.</p>
<h4>Netdata</h4>
<p>Netdata<sup id="fnref:Netdata"><a class="footnote-ref" href="../2024/06/macos-log-track-cpu-ram-usage.html#fn:Netdata">1</a></sup> is an open source observability tool that can monitor CPU, RAM, network, disk usage. It can also track the historical data. </p>
<p>Unfortunately, it is not stable on macOS. I tried installing it on multiple macbooks, but it didn't work. I raised an issue<sup id="fnref:netdata_issue"><a class="footnote-ref" href="../2024/06/macos-log-track-cpu-ram-usage.html#fn:netdata_issue">2</a></sup> on their GitHub repository and the team mentioned that macOS is a low priority for them.</p>
<h4>Glances</h4>
<p>Glances<sup id="fnref:Glances"><a class="footnote-ref" href="../2024/06/macos-log-track-cpu-ram-usage.html#fn:Glances">3</a></sup> is a cross-platform monitoring tool that can monitor CPU, RAM, network, disk usage. It can also track the historical data.</p>
<p>We can install it using Brew or pip.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>brew<span class="w"> </span>install<span class="w"> </span>glances

$<span class="w"> </span>pip<span class="w"> </span>install<span class="w"> </span>glances
</pre></div>

<p>Once it is installed, we can monitor the resource usage using the below command.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>glances
</pre></div>

<p><img alt="macOS - Log CPU &amp; RAM history" src="../images/mac-log-cpu-ram-glances.png"></p>
<p>Glances can log historical data to a file using the below command.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>glances<span class="w"> </span>--export-csv<span class="w"> </span>/tmp/glances.csv
</pre></div>

<p>In addition to that, it can log data to services like influxdb, prometheus, etc.</p>
<p>Let's install influxdb and export stats to it.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>brew<span class="w"> </span>install<span class="w"> </span>influxdb
$<span class="w"> </span>brew<span class="w"> </span>services<span class="w"> </span>start<span class="w"> </span>influxdb
$<span class="w"> </span>influx<span class="w"> </span>setup

$<span class="w"> </span>python<span class="w"> </span>-m<span class="w"> </span>pip<span class="w"> </span>install<span class="w"> </span>influxdb-client

$<span class="w"> </span>cat<span class="w"> </span>glances.conf
<span class="o">[</span>influxdb<span class="o">]</span>
<span class="nv">host</span><span class="o">=</span>localhost
<span class="nv">port</span><span class="o">=</span><span class="m">8086</span>
<span class="nv">protocol</span><span class="o">=</span>http
<span class="nv">org</span><span class="o">=</span>avilpage
<span class="nv">bucket</span><span class="o">=</span>glances
<span class="nv">token</span><span class="o">=</span>secret_token

$<span class="w"> </span>glances<span class="w"> </span>--export-influxdb<span class="w"> </span>-C<span class="w"> </span>glances.conf
</pre></div>

<p>We can view stats in the influxdb from Data Explorer web UI at <a href="http://localhost:8086">http://localhost:8086</a>.</p>
<p><img alt="macOS - Log CPU &amp; RAM history" src="../images/mac-log-cpu-ram-influxdb.png"></p>
<p>Glances provides a prebuilt Grafana dashboard<sup id="fnref:grafana_dashboard"><a class="footnote-ref" href="../2024/06/macos-log-track-cpu-ram-usage.html#fn:grafana_dashboard">4</a></sup> that we can import to visualize the stats. </p>
<p>From Grafana -&gt; Dashboard -&gt; Import, we can import the dashboard using the above URL.</p>
<p><img alt="macOS - Log CPU &amp; RAM history" src="../images/mac-log-cpu-ram-grafana.png"></p>
<h4>Conclusion</h4>
<p>In addition to InfluxDB, Glances can export data to ~20 services. So far, it is the best tool to log, track and view historical CPU, RAM, network and disk usage in macOS. The same method works for Linux and Windows as well.</p>
<div class="footnote">
<hr>
<ol>
<li id="fn:Netdata">
<p><a href="https://github.com/netdata/netdata">https://github.com/netdata/netdata</a> <a class="footnote-backref" href="../2024/06/macos-log-track-cpu-ram-usage.html#fnref:Netdata" title="Jump back to footnote 1 in the text">↩</a></p>
</li>
<li id="fn:netdata_issue">
<p><a href="https://github.com/netdata/netdata/issues/16696">https://github.com/netdata/netdata/issues/16696</a> <a class="footnote-backref" href="../2024/06/macos-log-track-cpu-ram-usage.html#fnref:netdata_issue" title="Jump back to footnote 2 in the text">↩</a></p>
</li>
<li id="fn:Glances">
<p><a href="https://github.com/nicolargo/glances">https://github.com/niolargo/glances</a> <a class="footnote-backref" href="../2024/06/macos-log-track-cpu-ram-usage.html#fnref:Glances" title="Jump back to footnote 3 in the text">↩</a></p>
</li>
<li id="fn:grafana_dashboard">
<p><a href="https://glances.readthedocs.io/en/latest/gw/influxdb.html#grafana">https://glances.readthedocs.io/en/latest/gw/influxdb.html#grafana</a> <a class="footnote-backref" href="../2024/06/macos-log-track-cpu-ram-usage.html#fnref:grafana_dashboard" title="Jump back to footnote 4 in the text">↩</a></p>
</li>
</ol>
</div>
    </div>
    </article><article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="../2024/05/managing-zscaler-connectivity-from-command-line.html" class="u-url">Automating Zscaler Connectivity on macOS</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                Chillar Anand
            </span></p>
            <p class="dateline">
            <a href="../2024/05/managing-zscaler-connectivity-from-command-line.html" rel="bookmark">
            <time class="published dt-published" datetime="2024-05-14T07:49:53+05:30" itemprop="datePublished" title="2024-05-14">2024-05-14</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <h4>Introduction</h4>
<p>Zscaler is a cloud-based security service that provides secure internet access via VPN. Unfortunately, Zscaler does not provide a command-line interface to connect to the VPN. We can't use AppleScript to automate the connectivity as well.</p>
<h4>Automating Zscaler Connectivity</h4>
<p>Once Zscaler is installed on macOS, if we search for <code>LaunchAgents</code> &amp; <code>LaunchDaemons</code> directories, we can find the Zscaler plist files. </p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>sudo<span class="w"> </span>find<span class="w"> </span>/Library/LaunchAgents<span class="w"> </span>-name<span class="w"> </span><span class="s1">'*zscaler*'</span>
/Library/LaunchAgents/com.zscaler.tray.plist


$<span class="w"> </span>sudo<span class="w"> </span>find<span class="w"> </span>/Library/LaunchDaemons<span class="w"> </span>-name<span class="w"> </span><span class="s1">'*zscaler*'</span>
/Library/LaunchDaemons/com.zscaler.service.plist
/Library/LaunchDaemons/com.zscaler.tunnel.plist
/Library/LaunchDaemons/com.zscaler.UPMServiceController.plist
</pre></div>

<p>To connect to Zscaler, we can load these services.</p>
<div class="code"><pre class="code literal-block"><span class="ch">#!/bin/bash</span>

/usr/bin/open<span class="w"> </span>-a<span class="w"> </span>/Applications/Zscaler/Zscaler.app<span class="w"> </span>--hide
sudo<span class="w"> </span>find<span class="w"> </span>/Library/LaunchAgents<span class="w"> </span>-name<span class="w"> </span><span class="s1">'*zscaler*'</span><span class="w"> </span>-exec<span class="w"> </span>launchctl<span class="w"> </span>load<span class="w"> </span><span class="o">{}</span><span class="w"> </span><span class="se">\;</span>
sudo<span class="w"> </span>find<span class="w"> </span>/Library/LaunchDaemons<span class="w"> </span>-name<span class="w"> </span><span class="s1">'*zscaler*'</span><span class="w"> </span>-exec<span class="w"> </span>launchctl<span class="w"> </span>load<span class="w"> </span><span class="o">{}</span><span class="w"> </span><span class="se">\;</span>
</pre></div>

<p>To disconnect from Zscaler, we can unload all of them.</p>
<div class="code"><pre class="code literal-block"><span class="ch">#!/bin/bash</span>

sudo<span class="w"> </span>find<span class="w"> </span>/Library/LaunchAgents<span class="w"> </span>-name<span class="w"> </span><span class="s1">'*zscaler*'</span><span class="w"> </span>-exec<span class="w"> </span>launchctl<span class="w"> </span>unload<span class="w"> </span><span class="o">{}</span><span class="w"> </span><span class="se">\;</span>
sudo<span class="w"> </span>find<span class="w"> </span>/Library/LaunchDaemons<span class="w"> </span>-name<span class="w"> </span><span class="s1">'*zscaler*'</span><span class="w"> </span>-exec<span class="w"> </span>launchctl<span class="w"> </span>unload<span class="w"> </span><span class="o">{}</span><span class="w"> </span><span class="se">\;</span>
</pre></div>

<p>To automatically toggle the connectivity, we can create a shell script.</p>
<div class="code"><pre class="code literal-block"><span class="ch">#!/bin/bash</span>

<span class="k">if</span><span class="w"> </span><span class="o">[[</span><span class="w"> </span><span class="k">$(</span>pgrep<span class="w"> </span>-x<span class="w"> </span>Zscaler<span class="k">)</span><span class="w"> </span><span class="o">]]</span><span class="p">;</span><span class="w"> </span><span class="k">then</span>
<span class="w">    </span><span class="nb">echo</span><span class="w"> </span><span class="s2">"Disconnecting from Zscaler"</span>
<span class="w">    </span>sudo<span class="w"> </span>find<span class="w"> </span>/Library/LaunchAgents<span class="w"> </span>-name<span class="w"> </span><span class="s1">'*zscaler*'</span><span class="w"> </span>-exec<span class="w"> </span>launchctl<span class="w"> </span>unload<span class="w"> </span><span class="o">{}</span><span class="w"> </span><span class="se">\;</span>
<span class="w">    </span>sudo<span class="w"> </span>find<span class="w"> </span>/Library/LaunchDaemons<span class="w"> </span>-name<span class="w"> </span><span class="s1">'*zscaler*'</span><span class="w"> </span>-exec<span class="w"> </span>launchctl<span class="w"> </span>unload<span class="w"> </span><span class="o">{}</span><span class="w"> </span><span class="se">\;</span>
<span class="k">else</span>
<span class="w">    </span><span class="nb">echo</span><span class="w"> </span><span class="s2">"Connecting to Zscaler"</span>
<span class="w">    </span>/usr/bin/open<span class="w"> </span>-a<span class="w"> </span>/Applications/Zscaler/Zscaler.app<span class="w"> </span>--hide
<span class="w">    </span>sudo<span class="w"> </span>find<span class="w"> </span>/Library/LaunchAgents<span class="w"> </span>-name<span class="w"> </span><span class="s1">'*zscaler*'</span><span class="w"> </span>-exec<span class="w"> </span>launchctl<span class="w"> </span>load<span class="w"> </span><span class="o">{}</span><span class="w"> </span><span class="se">\;</span>
<span class="w">    </span>sudo<span class="w"> </span>find<span class="w"> </span>/Library/LaunchDaemons<span class="w"> </span>-name<span class="w"> </span><span class="s1">'*zscaler*'</span><span class="w"> </span>-exec<span class="w"> </span>launchctl<span class="w"> </span>load<span class="w"> </span><span class="o">{}</span><span class="w"> </span><span class="se">\;</span>
<span class="k">fi</span>
</pre></div>

<p>Raycast is an alternative to default spotlight search on macOS. We can create a script to toggle connectivity to Zscaler.</p>
<div class="code"><pre class="code literal-block"><span class="ch">#!/bin/bash</span>

<span class="c1"># Required parameters:</span>
<span class="c1"># @raycast.schemaVersion 1</span>
<span class="c1"># @raycast.title toggle zscaler</span>
<span class="c1"># @raycast.mode silent</span>

<span class="c1"># Optional parameters:</span>
<span class="c1"># @raycast.icon ☁️</span>

<span class="c1"># Documentation:</span>
<span class="c1"># @raycast.author chillaranand</span>
<span class="c1"># @raycast.authorURL https://avilpage.com/</span>

<span class="k">if</span><span class="w"> </span><span class="o">[[</span><span class="w"> </span><span class="k">$(</span>pgrep<span class="w"> </span>-x<span class="w"> </span>Zscaler<span class="k">)</span><span class="w"> </span><span class="o">]]</span><span class="p">;</span><span class="w"> </span><span class="k">then</span>
<span class="w">    </span><span class="nb">echo</span><span class="w"> </span><span class="s2">"Disconnecting from Zscaler"</span>
<span class="w">    </span>sudo<span class="w"> </span>find<span class="w"> </span>/Library/LaunchAgents<span class="w"> </span>-name<span class="w"> </span><span class="s1">'*zscaler*'</span><span class="w"> </span>-exec<span class="w"> </span>launchctl<span class="w"> </span>unload<span class="w"> </span><span class="o">{}</span><span class="w"> </span><span class="se">\;</span>
<span class="w">    </span>sudo<span class="w"> </span>find<span class="w"> </span>/Library/LaunchDaemons<span class="w"> </span>-name<span class="w"> </span><span class="s1">'*zscaler*'</span><span class="w"> </span>-exec<span class="w"> </span>launchctl<span class="w"> </span>unload<span class="w"> </span><span class="o">{}</span><span class="w"> </span><span class="se">\;</span>
<span class="k">else</span>
<span class="w">    </span><span class="nb">echo</span><span class="w"> </span><span class="s2">"Connecting to Zscaler"</span>
<span class="w">    </span>/usr/bin/open<span class="w"> </span>-a<span class="w"> </span>/Applications/Zscaler/Zscaler.app<span class="w"> </span>--hide
<span class="w">    </span>sudo<span class="w"> </span>find<span class="w"> </span>/Library/LaunchAgents<span class="w"> </span>-name<span class="w"> </span><span class="s1">'*zscaler*'</span><span class="w"> </span>-exec<span class="w"> </span>launchctl<span class="w"> </span>load<span class="w"> </span><span class="o">{}</span><span class="w"> </span><span class="se">\;</span>
<span class="w">    </span>sudo<span class="w"> </span>find<span class="w"> </span>/Library/LaunchDaemons<span class="w"> </span>-name<span class="w"> </span><span class="s1">'*zscaler*'</span><span class="w"> </span>-exec<span class="w"> </span>launchctl<span class="w"> </span>load<span class="w"> </span><span class="o">{}</span><span class="w"> </span><span class="se">\;</span>
<span class="k">fi</span>
</pre></div>

<p>Save this script to a folder. From <code>Raycast Settings</code> -&gt; <code>Extensions</code> -&gt; <code>Add Script Directory</code>, we can select this folder, and the script will be available in Raycast. </p>
<p><img alt="raycast-connect-toggle" src="../images/automate-zscaler-connect2.png"></p>
<p>We can assign a shortcut key to the script for quick access.</p>
<p><img alt="raycast-connect-toggle" src="../images/automate-zscaler-connect.png"></p>
<h4>Conclusion</h4>
<p>Even though Zscaler does not provide a command-line interface, we can automate the connectivity using the above scripts.</p>
    </div>
    </article><article class="h-entry post-text" itemscope="itemscope" itemtype="http://schema.org/Article"><header><h1 class="p-name entry-title"><a href="../2024/05/screen-time-alerts-from-activity-watch.html" class="u-url">Screen Time Alerts from Activity Watch</a></h1>
        <div class="metadata">
            <p class="byline author vcard"><span class="byline-name fn" itemprop="author">
                Chillar Anand
            </span></p>
            <p class="dateline">
            <a href="../2024/05/screen-time-alerts-from-activity-watch.html" rel="bookmark">
            <time class="published dt-published" datetime="2024-05-01T06:59:27+05:30" itemprop="datePublished" title="2024-05-01">2024-05-01</time></a>
            </p>
        </div>
    </header><div class="e-content entry-content">
    <h4>Introduction</h4>
<div class="embed-responsive embed-responsive-16by9">
<iframe class="embed-responsive-item" src="https://www.youtube.com/embed/41A0r2H-EXs" allowfullscreen>
</iframe>
</div>

<p><br></p>
<p>Activity Watch<sup id="fnref:activity_watch"><a class="footnote-ref" href="../2024/05/screen-time-alerts-from-activity-watch.html#fn:activity_watch">1</a></sup> is a cross-platform open-source time-tracking tool that helps us to track time spent on applications and websites.</p>
<p><img alt="Activity Watch" src="../images/activity-watch-alerts.png"></p>
<p>At the moment, Activity Watch doesn't have any feature to show screen time alerts. In this post, we will see how to show screen time alerts using Activity Watch.</p>
<h4>Python Script</h4>
<p>Activity Watch provides an API to interact with the Activity Watch server. We can use the API to get the screen time data and show alerts.</p>
<div class="code"><pre class="code literal-block"><span class="kn">import</span><span class="w"> </span><span class="nn">json</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">datetime</span><span class="w"> </span><span class="kn">import</span> <span class="n">datetime</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">requests</span>


<span class="k">def</span><span class="w"> </span><span class="nf">get_nonafk_events</span><span class="p">(</span><span class="n">timeperiods</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="n">headers</span> <span class="o">=</span> <span class="p">{</span><span class="s2">"Content-type"</span><span class="p">:</span> <span class="s2">"application/json"</span><span class="p">,</span> <span class="s2">"charset"</span><span class="p">:</span> <span class="s2">"utf-8"</span><span class="p">}</span>
    <span class="n">query</span> <span class="o">=</span> <span class="s2">"""afk_events = query_bucket(find_bucket('aw-watcher-afk_'));</span>
<span class="s2">window_events = query_bucket(find_bucket('aw-watcher-window_'));</span>
<span class="s2">window_events = filter_period_intersect(window_events, filter_keyvals(afk_events, 'status', ['not-afk']));</span>
<span class="s2">RETURN = merge_events_by_keys(window_events, ['app', 'title']);"""</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">"</span><span class="se">\n</span><span class="s2">"</span><span class="p">)</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">"timeperiods"</span><span class="p">:</span> <span class="n">timeperiods</span><span class="p">,</span>
        <span class="s2">"query"</span><span class="p">:</span> <span class="n">query</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">r</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">post</span><span class="p">(</span>
        <span class="s2">"http://localhost:5600/api/0/query/"</span><span class="p">,</span>
        <span class="n">data</span><span class="o">=</span><span class="nb">bytes</span><span class="p">(</span><span class="n">json</span><span class="o">.</span><span class="n">dumps</span><span class="p">(</span><span class="n">data</span><span class="p">),</span> <span class="s2">"utf-8"</span><span class="p">),</span>
        <span class="n">headers</span><span class="o">=</span><span class="n">headers</span><span class="p">,</span>
        <span class="n">params</span><span class="o">=</span><span class="p">{},</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">json</span><span class="o">.</span><span class="n">loads</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">text</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>


<span class="k">def</span><span class="w"> </span><span class="nf">main</span><span class="p">():</span>
    <span class="n">now</span> <span class="o">=</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span>
    <span class="n">timeperiods</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">"/"</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">now</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">hour</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">minute</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">second</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">isoformat</span><span class="p">(),</span> <span class="n">now</span><span class="o">.</span><span class="n">isoformat</span><span class="p">()])</span>
    <span class="p">]</span>
    <span class="n">events</span> <span class="o">=</span> <span class="n">get_nonafk_events</span><span class="p">(</span><span class="n">timeperiods</span><span class="p">)</span>

    <span class="n">total_time_secs</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">([</span><span class="n">event</span><span class="p">[</span><span class="s2">"duration"</span><span class="p">]</span> <span class="k">for</span> <span class="n">event</span> <span class="ow">in</span> <span class="n">events</span><span class="p">])</span>
    <span class="n">total_time_mins</span> <span class="o">=</span> <span class="n">total_time_secs</span> <span class="o">/</span> <span class="mi">60</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Total time: </span><span class="si">{</span><span class="n">total_time_mins</span><span class="si">}</span><span class="s2"> seconds"</span><span class="p">)</span>
    <span class="n">hours</span><span class="p">,</span> <span class="n">minutes</span> <span class="o">=</span> <span class="nb">divmod</span><span class="p">(</span><span class="n">total_time_mins</span><span class="p">,</span> <span class="mi">60</span><span class="p">)</span>
    <span class="n">minutes</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">minutes</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">"Screen Time: </span><span class="si">{</span><span class="n">hours</span><span class="si">}</span><span class="s2"> hours </span><span class="si">{</span><span class="n">minutes</span><span class="si">}</span><span class="s2"> minutes"</span><span class="p">)</span>

    <span class="c1"># show mac notification</span>
    <span class="n">os</span><span class="o">.</span><span class="n">system</span><span class="p">(</span><span class="sa">f</span><span class="s2">"osascript -e 'display notification </span><span class="se">\"</span><span class="si">{</span><span class="n">hours</span><span class="si">}</span><span class="s2"> hours </span><span class="si">{</span><span class="n">minutes</span><span class="si">}</span><span class="s2"> minutes</span><span class="se">\"</span><span class="s2"> with title </span><span class="se">\"</span><span class="s2">Screen TIme</span><span class="se">\"</span><span class="s2">'"</span><span class="p">)</span>


<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s2">"__main__"</span><span class="p">:</span>
    <span class="n">main</span><span class="p">()</span>
</pre></div>

<p>This script<sup id="fnref:github"><a class="footnote-ref" href="../2024/05/screen-time-alerts-from-activity-watch.html#fn:github">2</a></sup> will show the screen time alerts using the Activity Watch API. We can run this script using the below command.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>python<span class="w"> </span>screen_time_alerts.py
</pre></div>

<p><img alt="Screen Time Alerts" src="../images/activity-watch-alerts2.png"></p>
<p>We can set up a cron job to run this script every hour to show screen time alerts.</p>
<div class="code"><pre class="code literal-block">$<span class="w"> </span>crontab<span class="w"> </span>-e
<span class="m">0</span><span class="w"> </span>*<span class="w"> </span>*<span class="w"> </span>*<span class="w"> </span>*<span class="w"> </span>python<span class="w"> </span>screen_time_alerts.py
</pre></div>

<p>We can also modify the script to show alerts only when the screen time exceeds a certain limit.</p>
<h4>Conclusion</h4>
<p>Since Actvity Watch is open-source and provides an API, we can extend its functionality to show screen time alerts. We can also use the API to create custom reports and dashboards.</p>
<div class="footnote">
<hr>
<ol>
<li id="fn:activity_watch">
<p><a href="https://activitywatch.net/">Activity Watch</a> <a class="footnote-backref" href="../2024/05/screen-time-alerts-from-activity-watch.html#fnref:activity_watch" title="Jump back to footnote 1 in the text">↩</a></p>
</li>
<li id="fn:github">
<p><a href="https://github.com/ChillarAnand/avilpage.com/blob/master/scripts/activity_watch_alerts.py">Screen Time Alerts Script</a> <a class="footnote-backref" href="../2024/05/screen-time-alerts-from-activity-watch.html#fnref:github" title="Jump back to footnote 2 in the text">↩</a></p>
</li>
</ol>
</div>
    </div>
    </article>
</div>
        <ul class="pager postindexpager clearfix">
<li class="previous"><a href="." rel="prev">Newer posts</a></li>
            <li class="next"><a href="index-19.html" rel="next">Older posts</a></li>
        </ul>
<!--End of body content--><footer id="footer"><div class="full-article-footer">
    <div class="article-footer">

        <div class="avatar-module">
            <img class="avatar" height="100px" src="../images/chillaranand.jpg">
</div>

        <p class="avatar-module">
            <b>Chillar Anand</b>
            <br>
            Improving Health &amp; Wealth with Technology
            <br></p>

    </div>
</div>

<div class="container align-items-center justify-content-center d-flex">

<footer class="footer"><a href="https://github.com/ChillarAnand">
<img src="../images/icons8-github.svg" alt="github-chillar-anand" height="34"></a>

  

<a href="https://stackoverflow.com/users/2698552/chillar-anand">
<img src="../images/icons8-so.svg" alt="github-chillar-anand" height="30"></a>

  

<a href="https://youtube.com/@avilpage">
<img src="../images/icons8-youtube.svg" alt="github-chillar-anand" height="34"></a>

  

<a href="https://linkedin.com/in/chillaranand">
<img src="../images/icons8-linkedin.svg" alt="github-chillar-anand" height="34"></a>

  

<a href="https://twitter.com/chillaranand">
<img src="../images/icons8-twitter.svg" alt="github-chillar-anand" height="34"></a>

</footer>
</div>

<br><script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-8127755709093749" crossorigin="anonymous"></script></footer>
</div>
    </div>
    </div>

                <script src="../assets/js/jquery.min.js"></script><script src="../assets/js/popper.min.js"></script><script src="../assets/js/bootstrap.min.js"></script><script src="../assets/js/baguetteBox.min.js"></script><script>
    baguetteBox.run('div#content', {
        ignoreClass: 'islink',
        captions: function(element){var i=element.getElementsByTagName('img')[0];return i===undefined?'':i.alt;}});
    </script>
</body>
</html>
